// igniteui-react-gauges/esm2015/lib/BulletGraphVisualData.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { LinearGraphVisualData } from "./LinearGraphVisualData";
import { StringBuilder } from "igniteui-react-core";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let BulletGraphVisualData = /*@__PURE__*/ (() => {
    class BulletGraphVisualData extends LinearGraphVisualData {
        constructor() {
            super(...arguments);
            this._targetValuePath = null;
            this._valuePath = null;
            this._scaleBackgroundPath = null;
            this._targetValue = null;
            this._value = null;
        }
        get targetValuePath() {
            return this._targetValuePath;
        }
        set targetValuePath(a) {
            this._targetValuePath = a;
        }
        get valuePath() {
            return this._valuePath;
        }
        set valuePath(a) {
            this._valuePath = a;
        }
        get scaleBackgroundPath() {
            return this._scaleBackgroundPath;
        }
        set scaleBackgroundPath(a) {
            this._scaleBackgroundPath = a;
        }
        get targetValue() {
            return this._targetValue;
        }
        set targetValue(a) {
            this._targetValue = a;
        }
        get value() {
            return this._value;
        }
        set value(a) {
            this._value = a;
        }
        scaleByViewport() {
            super.scaleByViewport();
            if (this.targetValuePath != null) {
                this.targetValuePath.scaleByViewport(this.viewport);
            }
            if (this.valuePath != null) {
                this.valuePath.scaleByViewport(this.viewport);
            }
            if (this.scaleBackgroundPath != null) {
                this.scaleBackgroundPath.scaleByViewport(this.viewport);
            }
        }
        serialize() {
            let a = super.serialize();
            let b = new StringBuilder(0);
            let c = a.length == 0;
            if (this.targetValuePath != null) {
                if (c) {
                    c = false;
                }
                else {
                    b.l(", ");
                }
                b.l("targetValuePath: ");
                b.u(this.targetValuePath.serialize());
            }
            if (this.targetValue != null) {
                if (c) {
                    c = false;
                }
                else {
                    b.l(", ");
                }
                b.l("targetValue: ");
                b.u(this.targetValue.serialize());
            }
            if (this.valuePath != null) {
                if (c) {
                    c = false;
                }
                else {
                    b.l(", ");
                }
                b.l("valuePath: ");
                b.u(this.valuePath.serialize());
            }
            if (this.value != null) {
                if (c) {
                    c = false;
                }
                else {
                    b.l(", ");
                }
                b.l("value: ");
                b.u(this.value.serialize());
            }
            if (this.scaleBackgroundPath != null) {
                if (c) {
                    c = false;
                }
                else {
                    b.l(", ");
                }
                b.l("scaleBackgroundPath: ");
                b.u(this.scaleBackgroundPath.serialize());
            }
            if (c) {
                return a;
            }
            b.l("}");
            return a.substr(0, a.length - 1) + b.toString();
        }
    }
    BulletGraphVisualData.$t = /*@__PURE__*/ markType(BulletGraphVisualData, 'BulletGraphVisualData', LinearGraphVisualData.$);
    return BulletGraphVisualData;
})();
