// igniteui-react-gauges/esm2015/lib/FormatRadialGaugeLabelEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { EventArgs, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let FormatRadialGaugeLabelEventArgs = /*@__PURE__*/ (() => {
    class FormatRadialGaugeLabelEventArgs extends EventArgs {
        constructor() {
            super(...arguments);
            this.actualMinimumValue = 0;
            this.actualMaximumValue = 0;
            this.startAngle = 0;
            this.endAngle = 0;
            this.angle = 0;
            this.value = 0;
            this.label = null;
        }
    }
    FormatRadialGaugeLabelEventArgs.$t = /*@__PURE__*/ markType(FormatRadialGaugeLabelEventArgs, 'FormatRadialGaugeLabelEventArgs', EventArgs.$);
    return FormatRadialGaugeLabelEventArgs;
})();
