// igniteui-react-gauges/esm2015/lib/igr-radial-gauge.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as React from 'react';
import { ReactRenderer } from "igniteui-react-core";
import { TypeRegistrar } from "igniteui-react-core";
import { XamRadialGauge } from './XamRadialGauge';
import { CollectionAdapter, brushToString, stringToBrush, toPoint, fromPoint, fromBrushCollection, toBrushCollection, ensureBool, ensureEnum, initializePropertiesFromCss, NamePatcher, isValidProp, getModifiedProps, toSpinal, arrayFindByName } from "igniteui-react-core";
import { delegateCombine, delegateRemove } from "igniteui-react-core";
import { NotifyCollectionChangedAction } from "igniteui-react-core";
import { SweepDirection_$type } from "igniteui-react-core";
import { RadialGaugeNeedleShape_$type } from './RadialGaugeNeedleShape';
import { RadialGaugePivotShape_$type } from './RadialGaugePivotShape';
import { RadialGaugeScaleOversweepShape_$type } from './RadialGaugeScaleOversweepShape';
import { RadialGaugeBackingShape_$type } from './RadialGaugeBackingShape';
import { RadialGaugeDuplicateLabelOmissionStrategy_$type } from './RadialGaugeDuplicateLabelOmissionStrategy';
import { IgrFormatRadialGaugeLabelEventArgs } from './igr-format-radial-gauge-label-event-args';
import { IgrAlignRadialGaugeLabelEventArgs } from './igr-align-radial-gauge-label-event-args';
import { RadialGaugeStylingDefaults } from './RadialGaugeStylingDefaults';
//import { RadialGaugeRange } from './RadialGaugeRange';
import { XamRadialGaugeRange } from './XamRadialGaugeRange';
import { IgrRadialGaugeRangeCollection } from './igr-radial-gauge-range-collection';
import { ContentChildrenManager } from "igniteui-react-core";
import { IgrDoubleValueChangedEventArgs } from "igniteui-react-core";
export class IgrRadialGauge extends React.Component {
    constructor(props) {
        super(props);
        this.contentRanges = [];
        /**
        * The ranges actually present in the chart. Do not directly modify this array.
        * This array's contents can be modified by causing Angular to reproject the child content.
        * Or adding and removing ranges from the manual ranges collection on the ranges property.
        */
        this.actualRanges = [];
        this._ranges = null;
        this._rangesAdapter = null;
        this.__p = null;
        this._hasUserValues = new Set();
        this._stylingContainer = null;
        this._stylingParent = null;
        this._inStyling = false;
        this._formatLabel = null;
        this._formatLabel_wrapped = null;
        this._alignLabel = null;
        this._alignLabel_wrapped = null;
        this._valueChanged = null;
        this._valueChanged_wrapped = null;
        this._actualMinimumValueChange = null;
        this._actualMinimumValueChange_wrapped = null;
        this._actualMaximumValueChange = null;
        this._actualMaximumValueChange_wrapped = null;
        if (this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(this));
        }
        this._getMainRef = this._getMainRef.bind(this);
        this._contentChildrenManager = new ContentChildrenManager((ch) => ch.key || ch.props.name, (ch) => ch.key || ch.props.name, () => this._updateContentChildren());
        let container = null;
        if (document) {
            container = document.createElement("div");
            container.style.display = "block";
            container.style.width = "100%";
            container.style.height = "100%";
        }
        var root;
        root = container;
        this._renderer = new ReactRenderer(root, document, true, RadialGaugeStylingDefaults);
        this._implementation = this.createImplementation();
        this._container = this._renderer.getWrapper(container);
        //this._renderer.rootWrapper.append(this._container);
        //this._container.setStyleProperty("display", "block");
        //this._container.setStyleProperty("width", "100%");
        //this._container.setStyleProperty("height", "100%");
        this._wrapper = this._renderer;
        //var root: any;
        //root = this._container;
        //if (this._container.nativeElement != null) {
        //    root = this._container.nativeElement;
        //}
        var gauge = this.i;
        this._gauge = gauge;
        this._rangesAdapter = new CollectionAdapter(this.contentRanges, this.i.ranges, this.actualRanges, (c) => c.i, (i) => {
            i._provideRenderer(this._wrapper);
            if (this._container && this._container.getNativeElement().parentElement) {
                i._styling(this._container.getNativeElement(), this, this);
            }
        }, (i) => { i._provideRenderer(null); });
        gauge.provideContainer(this._renderer);
        this._renderer.addSizeWatcher(() => {
            this._gauge.containerResized();
        });
    }
    set height(value) {
        this._height = value;
        if (this._elRef) {
            this._elRef.style.height = value;
            this.containerResized();
        }
    }
    get height() {
        return this._height;
    }
    set width(value) {
        this._width = value;
        if (this._elRef) {
            this._elRef.style.width = value;
            this.containerResized();
        }
    }
    get width() {
        return this._width;
    }
    _updateContentChildren() {
        this.contentRanges.length = 0;
        let contentChildrenActual = this._contentChildrenManager.contentChildrenActual;
        for (let i = 0; i < contentChildrenActual.length; i++) {
            if ((XamRadialGaugeRange.$type).isAssignableFrom(contentChildrenActual[i].i.$type)) {
                this.contentRanges.push(contentChildrenActual[i]);
            }
        }
        if (this._rangesAdapter !== null) {
            this._rangesAdapter.notifyContentChanged();
        }
    }
    /**
     * A collection or manually added axes for the chart.
    */
    get ranges() {
        if (this._ranges === null) {
            let coll = new IgrRadialGaugeRangeCollection();
            let inner = coll._innerColl;
            inner.addListener((sender, e) => {
                switch (e.action) {
                    case NotifyCollectionChangedAction.Add:
                        this._rangesAdapter.insertManualItem(e.newStartingIndex, e.newItems.item(0));
                        break;
                    case NotifyCollectionChangedAction.Remove:
                        this._rangesAdapter.removeManualItemAt(e.oldStartingIndex);
                        break;
                    case NotifyCollectionChangedAction.Replace:
                        this._rangesAdapter.removeManualItemAt(e.oldStartingIndex);
                        this._rangesAdapter.insertManualItem(e.newStartingIndex, e.newItems.item(0));
                        break;
                    case NotifyCollectionChangedAction.Reset:
                        this._rangesAdapter.clearManualItems();
                        break;
                }
            });
            this._ranges = coll;
        }
        return this._ranges;
    }
    render() {
        // if (!this._childrenDiffer(this.props.children)) {
        // 	let div = React.createElement("div", {
        // 		ref: (ref) => {
        // 			this._elRef = ref;
        // 		},
        // 		children: this.props.children
        // 	});
        // 	return div;
        // } else {
        let children = this._contentChildrenManager.getChildren(this.props.children);
        let div = React.createElement("div", {
            className: "ig-radial-gauge igr-radial-gauge",
            ref: this._getMainRef,
            children: children
        });
        return div;
        //}
    }
    componentDidMount() {
        for (const p of Object.keys(this.props)) {
            if (isValidProp(this, p)) {
                {
                    this[p] = this.props[p];
                }
            }
        }
        this._elRef.style.width = this._width ? this._width : "";
        this._elRef.style.height = this._height ? this._height : "";
        this._elRef.appendChild(this._container.getNativeElement());
        this.containerResized();
        this.initializeContent();
    }
    shouldComponentUpdate(nextProps, nextState) {
        const mod = getModifiedProps(this.props, nextProps);
        for (const p of Object.keys(mod)) {
            if (isValidProp(this, p)) {
                this[p] = mod[p];
            }
        }
        return true;
    }
    _getMainRef(div) {
        this._elRef = div;
    }
    destroy() {
        this._gauge.destroy();
        this._wrapper.destroy();
    }
    createImplementation() {
        return new XamRadialGauge();
    }
    initializeContent() {
        this._styling(this._container.getNativeElement(), this);
        if (this.actualRanges && this.actualRanges.length > 0) {
            var currRange = this.actualRanges;
            for (var i = 0; i < currRange.length; i++) {
                currRange[i]._styling(this._container.getNativeElement(), this, this);
            }
        }
        this._updateContentChildren();
        this.i.containerResized();
    }
    componentWillUnmount() {
        //this._grid.destroy();
        this.destroy();
    }
    initializeProperties() {
        for (const p of Object.keys(this.props)) {
            if (isValidProp(this, p)) {
                this[p] = this.props[p];
            }
        }
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    } /**
                                 * @hidden
                                 */
    static _createFromInternal(internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        let name = internal.$type.name;
        let externalName = "Igr" + name;
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    }
    /**
 * Gets or sets a collection of brushes to be used as the palette for gauge ranges.
*/
    get rangeBrushes() {
        return fromBrushCollection(this.i.ah);
    }
    set rangeBrushes(v) {
        this.i.ah = toBrushCollection(v);
    }
    /**
     * Gets or sets a collection of brushes to be used as the palette for gauge outlines.
    */
    get rangeOutlines() {
        return fromBrushCollection(this.i.ai);
    }
    set rangeOutlines(v) {
        this.i.ai = toBrushCollection(v);
    }
    /**
     * Gets or sets the minimum value of the scale.
    */
    get minimumValue() {
        return this.i.cc;
    }
    set minimumValue(v) {
        this.i.cc = +v;
    }
    /**
     * Gets the resolved minimum value of the scale.
    */
    get actualMinimumValue() {
        return this.i.bs;
    }
    set actualMinimumValue(v) {
        this.i.bs = +v;
    }
    /**
     * Gets or sets the maximum value of the scale.
    */
    get maximumValue() {
        return this.i.ca;
    }
    set maximumValue(v) {
        this.i.ca = +v;
    }
    /**
     * Gets the resolved maximum value of the scale.
    */
    get actualMaximumValue() {
        return this.i.br;
    }
    set actualMaximumValue(v) {
        this.i.br = +v;
    }
    /**
     * Gets or sets the interval to use for the scale.
    */
    get interval() {
        return this.i.b6;
    }
    set interval(v) {
        this.i.b6 = +v;
    }
    /**
     * Gets or sets the x position of the center of the gauge with the value ranging from 0 to 1.
    */
    get centerX() {
        return this.i.b3;
    }
    set centerX(v) {
        this.i.b3 = +v;
    }
    /**
     * Gets or sets the y position of the center of the gauge with the value ranging from 0 to 1.
    */
    get centerY() {
        return this.i.b4;
    }
    set centerY(v) {
        this.i.b4 = +v;
    }
    /**
     * Gets or sets the value at which to point the needle of the gauge.
    */
    get value() {
        return this.i.dm;
    }
    set value(v) {
        this.i.dm = +v;
    }
    /**
     * Gets or sets the start angle for the scale in degrees.
    */
    get scaleStartAngle() {
        return this.i.da;
    }
    set scaleStartAngle(v) {
        this.i.da = +v;
    }
    /**
     * Gets or sets the end angle for the scale in degrees.
    */
    get scaleEndAngle() {
        return this.i.c5;
    }
    set scaleEndAngle(v) {
        this.i.c5 = +v;
    }
    /**
     * Gets or sets the direction in which the scale sweeps around the center from the start angle to end angle.
    */
    get scaleSweepDirection() {
        return this.i.ii;
    }
    set scaleSweepDirection(v) {
        this.i.ii = ensureEnum(SweepDirection_$type, v);
    }
    /**
     * Gets or sets the number of milliseconds over which changes to the gauge should be animated.
    */
    get transitionDuration() {
        return this.i.dn;
    }
    set transitionDuration(v) {
        this.i.dn = +v;
    }
    /**
     * Gets or sets the easing function used to morph the current series.
    */
    get transitionEasingFunction() {
        return this.i.a0;
    }
    set transitionEasingFunction(v) {
        this.i.a0 = v;
    }
    /**
     * Gets or sets the brush to use when rendering the fill of the needle.
    */
    get needleBrush() {
        return brushToString(this.i.ib);
    }
    set needleBrush(v) {
        this.i.ib = stringToBrush(v);
    }
    /**
     * Gets or sets the brush to use when rendering the outline of the needle.
    */
    get needleOutline() {
        return brushToString(this.i.ic);
    }
    set needleOutline(v) {
        this.i.ic = stringToBrush(v);
    }
    /**
     * Gets or sets the extent (from -1 to 1) at which to start rendering the needle, measured from the center of the gauge.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get needleStartExtent() {
        return this.i.cv;
    }
    set needleStartExtent(v) {
        this.i.cv = +v;
    }
    /**
     * Gets or sets the extent (from -1 to 1) at which to end rendering the needle, measured from the center of the gauge.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get needleEndExtent() {
        return this.i.cn;
    }
    set needleEndExtent(v) {
        this.i.cn = +v;
    }
    /**
     * Gets or sets the shape to use when rendering the needle from a number of options.
    */
    get needleShape() {
        return this.i.ar;
    }
    set needleShape(v) {
        this.i.ar = ensureEnum(RadialGaugeNeedleShape_$type, v);
    }
    /**
     * Gets or sets the width of the needle at its point using a value from (0 to 1). Note: Only some needle shapes respect this property.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get needleStartWidthRatio() {
        return this.i.cw;
    }
    set needleStartWidthRatio(v) {
        this.i.cw = +v;
    }
    /**
     * Gets or sets the width of the needle at its point using a value from (0 to 1). Note: Only some needle shapes respect this property.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get needleEndWidthRatio() {
        return this.i.co;
    }
    set needleEndWidthRatio(v) {
        this.i.co = +v;
    }
    /**
     * Gets or sets the width of the needle at its feature which is closest to the base (e.g. a bulb) with a value from 0 to 1. Note: Only some needle shapes respect this property, namely: NeedleWithBulb, RectangleWithBulb, TrapezoidWithBulb, and TriangleWithBulb.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get needleBaseFeatureWidthRatio() {
        return this.i.cm;
    }
    set needleBaseFeatureWidthRatio(v) {
        this.i.cm = +v;
    }
    /**
     * Gets or sets the extent of the feature which is closest to the base (e.g. a bulb) with a value from -1 to 1. Note: Only some needle shapes respect this property, namely: NeedleWithBulb, RectangleWithBulb, TrapezoidWithBulb, and TriangleWithBulb.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get needleBaseFeatureExtent() {
        return this.i.cl;
    }
    set needleBaseFeatureExtent(v) {
        this.i.cl = +v;
    }
    /**
     * Gets or sets the width of the needle at its feature which is closest to the point (e.g. the tapering point of a needle) with a value from 0 to 1. Note: Only some needle shapes respect this property.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get needlePointFeatureWidthRatio() {
        return this.i.cu;
    }
    set needlePointFeatureWidthRatio(v) {
        this.i.cu = +v;
    }
    /**
     * Gets or sets the extent of the feature which is closest to the point (e.g. the tapering point of a needle) with a value from -1 to 1. Note: Only some needle shapes respect this property.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get needlePointFeatureExtent() {
        return this.i.ct;
    }
    set needlePointFeatureExtent(v) {
        this.i.ct = +v;
    }
    /**
     * Gets or sets the width of the cap of the needle with a value from 0 to 1. Note: Will only take effect if you have a cap set on the needle.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get needlePivotWidthRatio() {
        return this.i.cs;
    }
    set needlePivotWidthRatio(v) {
        this.i.cs = +v;
    }
    /**
     * Gets or sets the width of the inner cutout section of the needle cap with a value from 0 to 1. Note: Will only take effect if you have a cap set on the needle that has a cutout section.
    */
    get needlePivotInnerWidthRatio() {
        return this.i.cp;
    }
    set needlePivotInnerWidthRatio(v) {
        this.i.cp = +v;
    }
    /**
     * Gets or sets the shape to use for the needle cap.
    */
    get needlePivotShape() {
        return this.i.as;
    }
    set needlePivotShape(v) {
        this.i.as = ensureEnum(RadialGaugePivotShape_$type, v);
    }
    /**
     * Gets or sets the position at which to start rendering the scale, measured from the center of the gauge as a value from 0 to 1.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get scaleStartExtent() {
        return this.i.db;
    }
    set scaleStartExtent(v) {
        this.i.db = +v;
    }
    /**
     * Gets or sets the brush to use for filling the needle cap. Note: this only applies to certain cap shapes.
    */
    get needlePivotBrush() {
        return brushToString(this.i.id);
    }
    set needlePivotBrush(v) {
        this.i.id = stringToBrush(v);
    }
    /**
     * Gets or sets the brush to use for the outlines of the needle cap.
    */
    get needlePivotOutline() {
        return brushToString(this.i.ie);
    }
    set needlePivotOutline(v) {
        this.i.ie = stringToBrush(v);
    }
    /**
     * Gets or sets the stroke thickness of the needle outline.
    */
    get needleStrokeThickness() {
        return this.i.cx;
    }
    set needleStrokeThickness(v) {
        this.i.cx = +v;
    }
    /**
     * Gets or sets the stroke thickness to use for the outline of the needle cap.
    */
    get needlePivotStrokeThickness() {
        return this.i.cq;
    }
    set needlePivotStrokeThickness(v) {
        this.i.cq = +v;
    }
    /**
     * Gets or sets the position at which to stop rendering the scale as a value from 0 to 1 measured from the center of the gauge.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get scaleEndExtent() {
        return this.i.c6;
    }
    set scaleEndExtent(v) {
        this.i.c6 = +v;
    }
    /**
     * Gets or sets the position at which to put the labels as a value from 0 to 1, measured form the center of the gauge.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get labelExtent() {
        return this.i.b7;
    }
    set labelExtent(v) {
        this.i.b7 = +v;
    }
    /**
     * Gets or sets the interval to use for rendering labels. This defaults to be the same interval as the tickmarks on the scale.
    */
    get labelInterval() {
        return this.i.b9;
    }
    set labelInterval(v) {
        this.i.b9 = +v;
    }
    /**
     * Gets or sets the position at which to start rendering the major tickmarks as a value from 0 to 1, measured from the center of the gauge.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get tickStartExtent() {
        return this.i.dg;
    }
    set tickStartExtent(v) {
        this.i.dg = +v;
    }
    /**
     * Gets or sets the position at which to stop rendering the major tickmarks as a value from 0 to 1, measured from the center of the gauge.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get tickEndExtent() {
        return this.i.de;
    }
    set tickEndExtent(v) {
        this.i.de = +v;
    }
    /**
     * Gets or sets the stroke thickness to use when rendering ticks.
    */
    get tickStrokeThickness() {
        return this.i.di;
    }
    set tickStrokeThickness(v) {
        this.i.di = +v;
    }
    /**
     * Gets or sets the brush to use for the major tickmarks.
    */
    get tickBrush() {
        return brushToString(this.i.ih);
    }
    set tickBrush(v) {
        this.i.ih = stringToBrush(v);
    }
    /**
     * Gets or sets the brush to use for the label font.
    */
    get fontBrush() {
        return brushToString(this.i.h8);
    }
    set fontBrush(v) {
        this.i.h8 = stringToBrush(v);
    }
    /**
     * Gets or sets the position at which to start rendering the minor tickmarks as a value from 0 to 1, measured from the center of the gauge.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get minorTickStartExtent() {
        return this.i.ch;
    }
    set minorTickStartExtent(v) {
        this.i.ch = +v;
    }
    /**
     * Gets or sets the position at which to stop rendering the minor tickmarks as a value from 0 to 1, measured from the center of the gauge.
     * Values further from zero than 1 can be used to make this extend further than the normal radius of the gauge.
    */
    get minorTickEndExtent() {
        return this.i.cf;
    }
    set minorTickEndExtent(v) {
        this.i.cf = +v;
    }
    /**
     * Gets or sets the stroke thickness to use when rendering minor ticks.
    */
    get minorTickStrokeThickness() {
        return this.i.cj;
    }
    set minorTickStrokeThickness(v) {
        this.i.cj = +v;
    }
    /**
     * Gets or sets the brush to use for the minor tickmarks.
    */
    get minorTickBrush() {
        return brushToString(this.i.ia);
    }
    set minorTickBrush(v) {
        this.i.ia = stringToBrush(v);
    }
    /**
     * Gets or sets the number of minor tickmarks to place between major tickmarks.
    */
    get minorTickCount() {
        return this.i.ce;
    }
    set minorTickCount(v) {
        this.i.ce = +v;
    }
    /**
     * Gets or sets the brush to use to fill the background of the scale.
    */
    get scaleBrush() {
        return brushToString(this.i.ig);
    }
    set scaleBrush(v) {
        this.i.ig = stringToBrush(v);
    }
    /**
     * Gets or sets the brush to use to fill the backing of the gauge.
    */
    get backingBrush() {
        return brushToString(this.i.h6);
    }
    set backingBrush(v) {
        this.i.h6 = stringToBrush(v);
    }
    /**
     * Gets or sets the brush to use for the outline of the backing.
    */
    get backingOutline() {
        return brushToString(this.i.h7);
    }
    set backingOutline(v) {
        this.i.h7 = stringToBrush(v);
    }
    /**
     * Gets or sets the stroke thickness of the backing outline.
    */
    get backingStrokeThickness() {
        return this.i.b1;
    }
    set backingStrokeThickness(v) {
        this.i.b1 = +v;
    }
    /**
     * Gets or sets the outer extent of the gauge backing.
    */
    get backingOuterExtent() {
        return this.i.by;
    }
    set backingOuterExtent(v) {
        this.i.by = +v;
    }
    /**
     * Gets or sets the over sweep angle to apply to the backing if it is displaying fitted (in degrees). Must be greater or equal to 0.
    */
    get backingOversweep() {
        return this.i.b0;
    }
    set backingOversweep(v) {
        this.i.b0 = +v;
    }
    /**
     * Gets or sets the extra degrees of sweep to apply to the scale background. Must be greater or equal to 0.
    */
    get scaleOversweep() {
        return this.i.c8;
    }
    set scaleOversweep(v) {
        this.i.c8 = +v;
    }
    /**
     * Gets or sets the oversweep shape to use for the excess fill area for the scale.
    */
    get scaleOversweepShape() {
        return this.i.au;
    }
    set scaleOversweepShape(v) {
        this.i.au = ensureEnum(RadialGaugeScaleOversweepShape_$type, v);
    }
    /**
     * Gets or sets the corner rounding radius to use for the fitted scale backings.
    */
    get backingCornerRadius() {
        return this.i.bu;
    }
    set backingCornerRadius(v) {
        this.i.bu = +v;
    }
    /**
     * Gets or sets the inner extent of the gauge backing.
    */
    get backingInnerExtent() {
        return this.i.bw;
    }
    set backingInnerExtent(v) {
        this.i.bw = +v;
    }
    /**
     * Gets or sets the type of shape to use for the backing of the gauge.
    */
    get backingShape() {
        return this.i.am;
    }
    set backingShape(v) {
        this.i.am = ensureEnum(RadialGaugeBackingShape_$type, v);
    }
    /**
     * Gets or sets the multiplying factor to apply to the normal radius of the gauge.
     * The radius of the gauge is defined by the minimum of the width and height of the control divided by 2.0.
     * This introduces a multiplicative factor to that value.
    */
    get radiusMultiplier() {
        return this.i.c0;
    }
    set radiusMultiplier(v) {
        this.i.c0 = +v;
    }
    /**
     * Gets or sets the strategy to use for omitting labels if the first and last label have the same value.
    */
    get duplicateLabelOmissionStrategy() {
        return this.i.an;
    }
    set duplicateLabelOmissionStrategy(v) {
        this.i.an = ensureEnum(RadialGaugeDuplicateLabelOmissionStrategy_$type, v);
    }
    /**
     * Gets or sets whether needle dragging is enabled or not.
    */
    get isNeedleDraggingEnabled() {
        return this.i.ba;
    }
    set isNeedleDraggingEnabled(v) {
        this.i.ba = ensureBool(v);
    }
    /**
     * Gets or sets whether the needle is constrained within the minimum and maximum value range during dragging.
    */
    get isNeedleDraggingConstrained() {
        return this.i.a9;
    }
    set isNeedleDraggingConstrained(v) {
        this.i.a9 = ensureBool(v);
    }
    /**
     * Gets or sets the font.
    */
    get font() {
        return this.i.d2;
    }
    set font(v) {
        this.i.d2 = v;
    }
    /**
     * Gets or sets the label composite format used when creating label values.
    */
    get labelFormat() {
        return this.i.d9;
    }
    set labelFormat(v) {
        this.i.d9 = v;
    }
    /**
     * Gets or sets the format specifiers to use with the LabelFormat string.
    */
    get labelFormatSpecifiers() {
        return this.i.ag;
    }
    set labelFormatSpecifiers(v) {
        this.i.ag = v;
    }
    /**
     * Gets a value indicating whether the radial gauge is currently animating.
    */
    get animating() {
        return this.i.a6;
    }
    /**
     * Gets the transition progress of the animation when the control is animating.
    */
    get transitionProgress() {
        return this.i.dk;
    }
    set transitionProgress(v) {
        this.i.dk = +v;
    }
    /**
     * Gets or sets the scaling value used to affect the pixel density of the control.
     * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
     * to appear blurry.
    */
    get pixelScalingRatio() {
        return this.i.cz;
    }
    set pixelScalingRatio(v) {
        this.i.cz = +v;
    }
    /**
     * Gets the actual pixel scaling ratio used to affect the pixel density of the control.
     * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
     * to appear blurry.
    */
    get actualPixelScalingRatio() {
        return this.i.bt;
    }
    set actualPixelScalingRatio(v) {
        this.i.bt = +v;
    }
    findByName(name) {
        if (this.ranges != null && this.ranges.findByName && this.ranges.findByName(name)) {
            return this.ranges.findByName(name);
        }
        if (this.labelFormatSpecifiers != null && arrayFindByName(this.labelFormatSpecifiers, name)) {
            return arrayFindByName(this.labelFormatSpecifiers, name);
        }
        return null;
    }
    get hasUserValues() {
        return this._hasUserValues;
    }
    __m(propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    }
    _styling(container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        let genericPrefix = "";
        let typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("RadialGauge");
        let additionalPrefixes = [];
        let prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        let b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            let basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            let parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            let parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    }
    /**
 * Scales a value on the gauge's main scale to an angle around the center point of the gauge, in radians.

* @param value  * The value to scale.
*/
    scaleValue(value) {
        let iv = this.i.dd(value);
        return (iv);
    }
    /**
     * Unscales a value from an angle in radians to the represented value along the main scale of the gauge.
    
    * @param angle  * The angle in radians to unscale to a value.
    */
    unscaleValue(angle) {
        let iv = this.i.dl(angle);
        return (iv);
    }
    /**
     * Gets the value for the main scale of the gauge for a given point within the bounds of the gauge.
    
    * @param point  * The point for which to retrieve the associated value.
    */
    getValueForPoint(point) {
        let iv = this.i.b5(toPoint(point));
        return (iv);
    }
    getPointForValue(value, extent) {
        let iv = this.i.ik(value, extent);
        return fromPoint(iv);
    }
    styleUpdated() {
        this.i.gd();
    }
    needleContainsPoint(point, isFinger) {
        let iv = this.i.needleContainsPoint(toPoint(point), isFinger);
        return (iv);
    }
    /**
     * Called by the UI framework to provide a UI container for rendering this control.
    
    * @param container  * The UI container element.
    */
    provideContainer(container) {
        this.i.provideContainer(container);
    }
    containerResized() {
        this.i.containerResized();
    }
    /**
     * Use to force the radial gauge to finish any deferred work before printing or evaluating its visual.
     * This should only be called if the visual of the radial gauge needs to be synchronously saved or evaluated.
     * Calling this method too often will hinder the performance of the radial gauge.
    
    */
    flush() {
        this.i.fs();
    }
    /**
     * Returns visuals as a serialized string.
    
    */
    exportSerializedVisualData() {
        let iv = this.i.d1();
        return (iv);
    }
    /**
     * Event which is raised when a label of the gauge is formatted.
     * Function takes first argument null and second argument ui.
     * Use ui.owner to obtain reference to gauge widget.
     * Use ui.actualMinimumValue to obtain the minimum value of gauge scale.
     * Use ui.actualMaximumValue to obtain the maximum value of gauge scale.
     * Use ui.startAngle to obtain the starting angle of gauge scale.
     * Use ui.endAngle to obtain the ending angle of gauge scale.
     * Use ui.angle to obtain the angle on the gauge scale at which the label will be located.
     * Use ui.value to obtain the value on the gauge scale associated with the label.
     * Use ui.label to obtain the string value of the label.
    */
    get formatLabel() {
        return this._formatLabel;
    }
    set formatLabel(ev) {
        if (this._formatLabel_wrapped !== null) {
            this.i.formatLabel = delegateRemove(this.i.formatLabel, this._formatLabel_wrapped);
            this._formatLabel_wrapped = null;
            this._formatLabel = null;
        }
        this._formatLabel = ev;
        this._formatLabel_wrapped = (o, e) => {
            let outerArgs = new IgrFormatRadialGaugeLabelEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeFormatLabel) {
                this.beforeFormatLabel(this, outerArgs);
            }
            if (this._formatLabel) {
                this._formatLabel(this, outerArgs);
            }
        };
        this.i.formatLabel = delegateCombine(this.i.formatLabel, this._formatLabel_wrapped);
    }
    /**
     * Event which is raised when a label of the gauge is aligned along the scale.
     * Function takes first argument null and second argument ui.
     * Use ui.owner to obtain reference to gauge widget.
     * Use ui.actualMinimumValue to obtain the minimum value of gauge scale.
     * Use ui.actualMaximumValue to obtain the maximum value of gauge scale.
     * Use ui.startAngle to obtain the starting angle of gauge scale.
     * Use ui.endAngle to obtain the ending angle of gauge scale.
     * Use ui.angle to obtain the angle on the gauge scale at which the label will be located.
     * Use ui.value to obtain the value on the gauge scale associated with the label.
     * Use ui.label to obtain the string value of the label.
     * Use ui.width to obtain the width of the label.
     * Use ui.height to obtain the height of the label.
     * Use ui.offsetX to obtain the X offset of the label on the gauge scale.
     * Use ui.offsetY to obtain the Y offset of the label on the gauge scale.
    */
    get alignLabel() {
        return this._alignLabel;
    }
    set alignLabel(ev) {
        if (this._alignLabel_wrapped !== null) {
            this.i.alignLabel = delegateRemove(this.i.alignLabel, this._alignLabel_wrapped);
            this._alignLabel_wrapped = null;
            this._alignLabel = null;
        }
        this._alignLabel = ev;
        this._alignLabel_wrapped = (o, e) => {
            let outerArgs = new IgrAlignRadialGaugeLabelEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeAlignLabel) {
                this.beforeAlignLabel(this, outerArgs);
            }
            if (this._alignLabel) {
                this._alignLabel(this, outerArgs);
            }
        };
        this.i.alignLabel = delegateCombine(this.i.alignLabel, this._alignLabel_wrapped);
    }
    /**
     * Occurs when the Value property changes.
    */
    get valueChanged() {
        return this._valueChanged;
    }
    set valueChanged(ev) {
        if (this._valueChanged_wrapped !== null) {
            this.i.valueChanged = delegateRemove(this.i.valueChanged, this._valueChanged_wrapped);
            this._valueChanged_wrapped = null;
            this._valueChanged = null;
        }
        this._valueChanged = ev;
        this._valueChanged_wrapped = (o, e) => {
            let outerArgs = new IgrDoubleValueChangedEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeValueChanged) {
                this.beforeValueChanged(this, outerArgs);
            }
            if (this._valueChanged) {
                this._valueChanged(this, outerArgs);
            }
        };
        this.i.valueChanged = delegateCombine(this.i.valueChanged, this._valueChanged_wrapped);
    }
    get actualMinimumValueChange() {
        return this._actualMinimumValueChange;
    }
    set actualMinimumValueChange(ev) {
        if (this._actualMinimumValueChange_wrapped !== null) {
            this.i.propertyChanged = delegateRemove(this.i.propertyChanged, this._actualMinimumValueChange_wrapped);
            this._actualMinimumValueChange_wrapped = null;
            this._actualMinimumValueChange = null;
        }
        this._actualMinimumValueChange = ev;
        this._actualMinimumValueChange_wrapped = (o, e) => {
            let ext = this.actualMinimumValue;
            if (this.beforeActualMinimumValueChange) {
                this.beforeActualMinimumValueChange(this, ext);
            }
            if (this._actualMinimumValueChange) {
                this._actualMinimumValueChange(this, ext);
            }
        };
        this.i.propertyChanged = delegateCombine(this.i.propertyChanged, this._actualMinimumValueChange_wrapped);
    }
    get actualMaximumValueChange() {
        return this._actualMaximumValueChange;
    }
    set actualMaximumValueChange(ev) {
        if (this._actualMaximumValueChange_wrapped !== null) {
            this.i.propertyChanged = delegateRemove(this.i.propertyChanged, this._actualMaximumValueChange_wrapped);
            this._actualMaximumValueChange_wrapped = null;
            this._actualMaximumValueChange = null;
        }
        this._actualMaximumValueChange = ev;
        this._actualMaximumValueChange_wrapped = (o, e) => {
            let ext = this.actualMaximumValue;
            if (this.beforeActualMaximumValueChange) {
                this.beforeActualMaximumValueChange(this, ext);
            }
            if (this._actualMaximumValueChange) {
                this._actualMaximumValueChange(this, ext);
            }
        };
        this.i.propertyChanged = delegateCombine(this.i.propertyChanged, this._actualMaximumValueChange_wrapped);
    }
}
