// igniteui-react-gauges/esm2015/lib/igr-align-linear-graph-label-event-args.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { IgrFormatLinearGraphLabelEventArgs } from "./igr-format-linear-graph-label-event-args";
import { AlignLinearGraphLabelEventArgs as AlignLinearGraphLabelEventArgs_internal } from "./AlignLinearGraphLabelEventArgs";
/**
 * Represents event argument for AlignLinearGraphLabel.
*/
export class IgrAlignLinearGraphLabelEventArgs extends IgrFormatLinearGraphLabelEventArgs {
    createImplementation() {
        return new AlignLinearGraphLabelEventArgs_internal();
    }
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    get width() {
        return this.i.width;
    }
    set width(value) {
        this.i.width = value;
    }
    get height() {
        return this.i.height;
    }
    set height(value) {
        this.i.height = value;
    }
    get offsetX() {
        return this.i.offsetX;
    }
    set offsetX(value) {
        this.i.offsetX = value;
    }
    get offsetY() {
        return this.i.offsetY;
    }
    set offsetY(value) {
        this.i.offsetY = value;
    }
}
