// igniteui-react-gauges/esm2015/lib/igr-format-radial-gauge-label-event-args.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FormatRadialGaugeLabelEventArgs as FormatRadialGaugeLabelEventArgs_internal } from "./FormatRadialGaugeLabelEventArgs";
/**
 * Represents event argument for FormatRadialGaugeLabel.
*/
export class IgrFormatRadialGaugeLabelEventArgs {
    createImplementation() {
        return new FormatRadialGaugeLabelEventArgs_internal();
    }
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    get actualMinimumValue() {
        return this.i.actualMinimumValue;
    }
    get actualMaximumValue() {
        return this.i.actualMaximumValue;
    }
    get startAngle() {
        return this.i.startAngle;
    }
    get endAngle() {
        return this.i.endAngle;
    }
    get angle() {
        return this.i.angle;
    }
    get value() {
        return this.i.value;
    }
    get label() {
        return this.i.label;
    }
    set label(value) {
        this.i.label = value;
    }
}
