// igniteui-react-gauges/esm2015/lib/LinearGraphNeedleShape.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Enumeration of preset needle shapes.
 */
export var LinearGraphNeedleShape = /*@__PURE__*/ (function (LinearGraphNeedleShape) {
    /**
     * A custom user defined needle shape.
     */
    LinearGraphNeedleShape[LinearGraphNeedleShape["Custom"] = 0] = "Custom";
    /**
     * A needle shaped like a rectangle.
     */
    LinearGraphNeedleShape[LinearGraphNeedleShape["Rectangle"] = 1] = "Rectangle";
    /**
     * A needle shaped like a triangle.
     */
    LinearGraphNeedleShape[LinearGraphNeedleShape["Triangle"] = 2] = "Triangle";
    /**
     * A needle shaped like a needle.
     */
    LinearGraphNeedleShape[LinearGraphNeedleShape["Needle"] = 3] = "Needle";
    /**
     * A needle shaped like a trapezoid.
     */
    LinearGraphNeedleShape[LinearGraphNeedleShape["Trapezoid"] = 4] = "Trapezoid";
    return LinearGraphNeedleShape;
})({});
/**
 * @hidden
 */
export let LinearGraphNeedleShape_$type = /*@__PURE__*/ markEnum('LinearGraphNeedleShape', 'Custom,0|Rectangle,1|Triangle,2|Needle,3|Trapezoid,4');
