// igniteui-react-gauges/esm2015/lib/LinearGraphNeedleFrame.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, Point_$type, markType } from "igniteui-react-core";
import { MorphSegment } from "./MorphSegment";
import { BrushUtil } from "igniteui-react-core";
import { PathGeometry } from "igniteui-react-core";
import { PathFigure } from "igniteui-react-core";
/**
 * @hidden
 */
export let LinearGraphNeedleFrame = /*@__PURE__*/ (() => {
    class LinearGraphNeedleFrame extends Base {
        constructor() {
            super();
            this.n = 0;
            this.m = 0;
            this.q = 0;
            this.p = 0;
            this.l = 0;
            this.r = null;
            this.s = null;
            this.o = 0;
            this.f = null;
            this.g = null;
            this.d = null;
            this.j = null;
            this.i = null;
            this.k = null;
            this.e = null;
            this.h = null;
            this.b = null;
            this.c = null;
            this.f = new MorphSegment();
            this.g = new MorphSegment();
            this.d = new MorphSegment();
            this.j = new MorphSegment();
            this.i = new MorphSegment();
            this.k = new MorphSegment();
            this.e = new MorphSegment();
            this.h = new MorphSegment();
            this.b = new MorphSegment();
            this.c = new MorphSegment();
        }
        static a(a, b, c, d) {
            let e = new LinearGraphNeedleFrame();
            e.n = c.n + (d.n - c.n) * a;
            e.m = c.m + (d.m - c.m) * a;
            e.q = c.q + (d.q - c.q) * a;
            e.l = c.l + (d.l - c.l) * a;
            e.p = c.p + (d.p - c.p) * a;
            e.r = BrushUtil.k(c.r, a, d.r, 0);
            e.s = BrushUtil.k(c.s, a, d.s, 0);
            e.o = c.o + (d.o - c.o) * a;
            e.f.j(a, c.f, d.f);
            e.g.j(a, c.g, d.g);
            e.d.j(a, c.d, d.d);
            e.j.j(a, c.j, d.j);
            e.i.j(a, c.i, d.i);
            e.k.j(a, c.k, d.k);
            e.e.j(a, c.e, d.e);
            e.h.j(a, c.h, d.h);
            e.b.j(a, c.b, d.b);
            e.c.j(a, c.c, d.c);
            return e;
        }
        t(a, b, c) {
            let d = new PathGeometry();
            let e = new PathFigure();
            let f = { $type: Point_$type, x: NaN, y: NaN };
            f = this.i.l(f, e, a, b, c);
            f = this.j.l(f, e, a, b, c);
            f = this.d.l(f, e, a, b, c);
            f = this.g.l(f, e, a, b, c);
            f = this.f.l(f, e, a, b, c);
            f = this.h.l(f, e, a, b, c);
            f = this.e.l(f, e, a, b, c);
            e._isClosed = true;
            d.c.add(e);
            let g = new PathFigure();
            if (this.b.f.count > 0 || this.c.f.count > 0) {
                f = this.b.l(f, g, a, b, c);
                f = this.c.l(f, g, a, b, c);
                g._isClosed = true;
                d.c.add(g);
            }
            return d;
        }
    }
    LinearGraphNeedleFrame.$t = /*@__PURE__*/ markType(LinearGraphNeedleFrame, 'LinearGraphNeedleFrame');
    return LinearGraphNeedleFrame;
})();
