// igniteui-react-gauges/esm2015/lib/RadialGaugePivotShape.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Enum specifying shapes for the needle pivot.
 */
export var RadialGaugePivotShape = /*@__PURE__*/ (function (RadialGaugePivotShape) {
    /**
     * No pivot shape.
     */
    RadialGaugePivotShape[RadialGaugePivotShape["None"] = 0] = "None";
    /**
     * A circle shaped pivot.
     */
    RadialGaugePivotShape[RadialGaugePivotShape["Circle"] = 1] = "Circle";
    /**
     * A circle pivot with a hole in it.
     */
    RadialGaugePivotShape[RadialGaugePivotShape["CircleWithHole"] = 2] = "CircleWithHole";
    /**
     * A circle pivot overlayed on top of the needle.
     */
    RadialGaugePivotShape[RadialGaugePivotShape["CircleOverlay"] = 3] = "CircleOverlay";
    /**
     * A circle pivot with a hole in it overlayed on top of the needle.
     */
    RadialGaugePivotShape[RadialGaugePivotShape["CircleOverlayWithHole"] = 4] = "CircleOverlayWithHole";
    /**
     * A circle pivot rendered underneath the needle.
     */
    RadialGaugePivotShape[RadialGaugePivotShape["CircleUnderlay"] = 5] = "CircleUnderlay";
    /**
     * A circle pivot with a hold in it rendered underneath the needle.
     */
    RadialGaugePivotShape[RadialGaugePivotShape["CircleUnderlayWithHole"] = 6] = "CircleUnderlayWithHole";
    return RadialGaugePivotShape;
})({});
/**
 * @hidden
 */
export let RadialGaugePivotShape_$type = /*@__PURE__*/ markEnum('RadialGaugePivotShape', 'None,0|Circle,1|CircleWithHole,2|CircleOverlay,3|CircleOverlayWithHole,4|CircleUnderlay,5|CircleUnderlayWithHole,6');
