// igniteui-react-gauges/esm2015/lib/RadialGaugeFrame.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { RadialGaugeNeedleFrame } from "./RadialGaugeNeedleFrame";
import { RadialGaugeScaleFrame } from "./RadialGaugeScaleFrame";
import { BackingFrame } from "./BackingFrame";
/**
 * @hidden
 */
export let RadialGaugeFrame = /*@__PURE__*/ (() => {
    class RadialGaugeFrame extends Base {
        constructor() {
            super();
            this.b = null;
            this.c = null;
            this.a = null;
            this.d = 0;
            this.b = new RadialGaugeNeedleFrame();
            this.c = new RadialGaugeScaleFrame();
            this.a = new BackingFrame();
        }
        e(a, b, c) {
            this.d = b.d + (c.d - b.d) * a;
            this.b.ag(a, b.b, c.b);
            this.c.x(a, b.c, c.c);
            this.a.k(a, b.a, c.a);
        }
    }
    RadialGaugeFrame.$t = /*@__PURE__*/ markType(RadialGaugeFrame, 'RadialGaugeFrame');
    return RadialGaugeFrame;
})();
