// igniteui-react-gauges/esm2015/lib/RadialGaugeNeedleShape.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
/**
 * Enum specifying various preset needle shapes.
 */
export var RadialGaugeNeedleShape = /*@__PURE__*/ (function (RadialGaugeNeedleShape) {
    /**
     * No shape.
     */
    RadialGaugeNeedleShape[RadialGaugeNeedleShape["None"] = 0] = "None";
    /**
     * A rectangle shape.
     */
    RadialGaugeNeedleShape[RadialGaugeNeedleShape["Rectangle"] = 1] = "Rectangle";
    /**
     * A triangle shape.
     */
    RadialGaugeNeedleShape[RadialGaugeNeedleShape["Triangle"] = 2] = "Triangle";
    /**
     * A needle shape.
     */
    RadialGaugeNeedleShape[RadialGaugeNeedleShape["Needle"] = 3] = "Needle";
    /**
     * A trapezoid shape.
     */
    RadialGaugeNeedleShape[RadialGaugeNeedleShape["Trapezoid"] = 4] = "Trapezoid";
    /**
     * A rectangle shape with a bulb at the end.
     */
    RadialGaugeNeedleShape[RadialGaugeNeedleShape["RectangleWithBulb"] = 5] = "RectangleWithBulb";
    /**
     * A triangle shape with a bulb at the end.
     */
    RadialGaugeNeedleShape[RadialGaugeNeedleShape["TriangleWithBulb"] = 6] = "TriangleWithBulb";
    /**
     * A needle shape with a bulb at the end.
     */
    RadialGaugeNeedleShape[RadialGaugeNeedleShape["NeedleWithBulb"] = 7] = "NeedleWithBulb";
    /**
     * A trapezoid shape with a bulb at the end.
     */
    RadialGaugeNeedleShape[RadialGaugeNeedleShape["TrapezoidWithBulb"] = 8] = "TrapezoidWithBulb";
    return RadialGaugeNeedleShape;
})({});
/**
 * @hidden
 */
export let RadialGaugeNeedleShape_$type = /*@__PURE__*/ markEnum('RadialGaugeNeedleShape', 'None,0|Rectangle,1|Triangle,2|Needle,3|Trapezoid,4|RectangleWithBulb,5|TriangleWithBulb,6|NeedleWithBulb,7|TrapezoidWithBulb,8');
