// igniteui-react-gauges/esm2015/lib/RadialGaugeNeedleParameters.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let RadialGaugeNeedleParameters = /*@__PURE__*/ (() => {
    class RadialGaugeNeedleParameters extends Base {
        constructor() {
            super(...arguments);
            this.m = 0;
            this.k = 0;
            this.n = 0;
            this.l = 0;
            this.e = 0;
            this.d = 0;
            this.q = 0;
            this.p = 0;
            this.j = 0;
            this.b = 0;
            this.c = 0;
            this.i = 0;
            this.r = null;
            this.s = null;
            this.h = 0;
            this.t = null;
            this.u = null;
            this.o = 0;
            this.g = 0;
            this.f = 0;
        }
        a() {
            let a = new RadialGaugeNeedleParameters();
            a.m = this.m;
            a.k = this.k;
            a.n = this.n;
            a.l = this.l;
            a.e = this.e;
            a.d = this.d;
            a.q = this.q;
            a.p = this.p;
            a.b = this.b;
            a.c = this.c;
            a.i = this.i;
            a.r = this.r;
            a.s = this.s;
            a.h = this.h;
            a.t = this.t;
            a.u = this.u;
            a.o = this.o;
            a.g = this.g;
            a.f = this.f;
            return a;
        }
    }
    RadialGaugeNeedleParameters.$t = /*@__PURE__*/ markType(RadialGaugeNeedleParameters, 'RadialGaugeNeedleParameters');
    return RadialGaugeNeedleParameters;
})();
