// igniteui-react-gauges/esm5/lib/LinearGraphNeedleParameters.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Base, INotifyPropertyChanged_$type, PropertyChangedEventArgs, markType } from "igniteui-react-core";
/**
 * @hidden
 */
var LinearGraphNeedleParameters = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(LinearGraphNeedleParameters, _super);
    function LinearGraphNeedleParameters() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.d = 0;
        _this.e = false;
        _this.k = NaN;
        _this.j = NaN;
        _this.g = NaN;
        _this.l = NaN;
        _this.i = NaN;
        _this.n = NaN;
        _this.h = NaN;
        _this.m = NaN;
        _this.f = NaN;
        _this.b = 3;
        _this.propertyChanged = null;
        return _this;
    }
    Object.defineProperty(LinearGraphNeedleParameters.prototype, "u", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            if (this.k != a) {
                this.k = a;
                this.x("OuterExtent");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LinearGraphNeedleParameters.prototype, "q", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            if (this.j != a) {
                this.j = a;
                this.x("InnerExtent");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LinearGraphNeedleParameters.prototype, "p", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            if (this.g != a) {
                this.g = a;
                this.x("InnerBaseWidth");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LinearGraphNeedleParameters.prototype, "t", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            if (this.l != a) {
                this.l = a;
                this.x("OuterBaseWidth");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LinearGraphNeedleParameters.prototype, "s", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            if (this.i != a) {
                this.i = a;
                this.x("InnerPointWidth");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LinearGraphNeedleParameters.prototype, "w", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            if (this.n != a) {
                this.n = a;
                this.x("OuterPointWidth");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LinearGraphNeedleParameters.prototype, "r", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            if (this.h != a) {
                this.h = a;
                this.x("InnerPointExtent");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LinearGraphNeedleParameters.prototype, "v", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            if (this.m != a) {
                this.m = a;
                this.x("OuterPointExtent");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LinearGraphNeedleParameters.prototype, "o", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            if (this.f != a) {
                this.f = a;
                this.x("DefaultWidth");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LinearGraphNeedleParameters.prototype, "c", {
        get: function () {
            return this.b;
        },
        set: function (a) {
            if (this.b != a) {
                this.b = a;
                this.x("NeedleShape");
            }
        },
        enumerable: true,
        configurable: true
    });
    LinearGraphNeedleParameters.prototype.a = function () {
        var a = new LinearGraphNeedleParameters();
        a.u = this.u;
        a.q = this.q;
        a.p = this.p;
        a.t = this.t;
        a.r = this.r;
        a.v = this.v;
        a.s = this.s;
        a.w = this.w;
        a.c = this.c;
        a.o = this.o;
        return a;
    };
    LinearGraphNeedleParameters.prototype.x = function (a) {
        if (this.propertyChanged != null) {
            this.propertyChanged(this, new PropertyChangedEventArgs(a));
        }
    };
    LinearGraphNeedleParameters.$t = markType(LinearGraphNeedleParameters, 'LinearGraphNeedleParameters', Base.$, [INotifyPropertyChanged_$type]);
    return LinearGraphNeedleParameters;
}(Base));
export { LinearGraphNeedleParameters };
