// igniteui-react-core/esm5/lib/ButtonGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ButtonGroupDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(ButtonGroupDescription, _super);
    function ButtonGroupDescription() {
        var _this = _super.call(this) || this;
        _this.as = null;
        _this.aq = null;
        _this.ar = null;
        _this.ad = null;
        _this.g = null;
        _this.o = 0;
        _this.ae = null;
        _this.ag = null;
        _this.al = null;
        _this.ai = null;
        _this.af = null;
        _this.ah = null;
        _this.n = 0;
        _this.am = null;
        _this.ap = null;
        _this.aj = null;
        _this.ak = null;
        _this.an = null;
        _this.ao = null;
        _this.s = 0;
        _this.a6 = null;
        _this.a8 = null;
        _this.bd = null;
        _this.ba = null;
        _this.bb = null;
        _this.a7 = null;
        _this.a9 = null;
        _this.r = 0;
        _this.bc = null;
        _this.br = null;
        _this.bu = null;
        _this.bs = null;
        _this.bt = null;
        _this.u = 0;
        _this.bf = null;
        _this.bh = null;
        _this.bm = null;
        _this.bj = null;
        _this.bk = null;
        _this.bg = null;
        _this.bi = null;
        _this.t = 0;
        _this.bl = null;
        _this.bn = null;
        _this.bq = null;
        _this.bo = null;
        _this.bp = null;
        _this.q = 0;
        _this.at = null;
        _this.av = null;
        _this.a0 = null;
        _this.ax = null;
        _this.ay = null;
        _this.au = null;
        _this.aw = null;
        _this.p = 0;
        _this.az = null;
        _this.a1 = null;
        _this.a4 = null;
        _this.a2 = null;
        _this.a3 = null;
        _this.be = null;
        _this.a5 = null;
        _this.j = false;
        _this.k = false;
        _this.f = null;
        _this.bv = null;
        return _this;
    }
    ButtonGroupDescription.prototype.get_type = function () {
        return "ButtonGroup";
    };
    Object.defineProperty(ButtonGroupDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "displayType", {
        get: function () {
            return this.as;
        },
        set: function (a) {
            this.as = a;
            this.e("DisplayType");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "baseTheme", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.e("BaseTheme");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "density", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.e("Density");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualDensity", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.e("ActualDensity");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "selectedIndices", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("SelectedIndices");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemCornerRadius", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.e("ActualItemCornerRadius");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemBackgroundColor", {
        get: function () {
            return this.ae;
        },
        set: function (a) {
            this.ae = a;
            this.e("ActualItemBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemDisabledBackgroundColor", {
        get: function () {
            return this.ag;
        },
        set: function (a) {
            this.ag = a;
            this.e("ActualItemDisabledBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemTextColor", {
        get: function () {
            return this.al;
        },
        set: function (a) {
            this.al = a;
            this.e("ActualItemTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemDisabledTextColor", {
        get: function () {
            return this.ai;
        },
        set: function (a) {
            this.ai = a;
            this.e("ActualItemDisabledTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemBorderColor", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            this.af = a;
            this.e("ActualItemBorderColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemDisabledBorderColor", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            this.ah = a;
            this.e("ActualItemDisabledBorderColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemBorderWidth", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.e("ActualItemBorderWidth");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualSelectedItemBackgroundColor", {
        get: function () {
            return this.am;
        },
        set: function (a) {
            this.am = a;
            this.e("ActualSelectedItemBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualSelectedItemTextColor", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            this.ap = a;
            this.e("ActualSelectedItemTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemHoverBackgroundColor", {
        get: function () {
            return this.aj;
        },
        set: function (a) {
            this.aj = a;
            this.e("ActualItemHoverBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualItemHoverTextColor", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            this.ak = a;
            this.e("ActualItemHoverTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualSelectedItemHoverBackgroundColor", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            this.an = a;
            this.e("ActualSelectedItemHoverBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "actualSelectedItemHoverTextColor", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            this.ao = a;
            this.e("ActualSelectedItemHoverTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemCornerRadius", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.e("ItemCornerRadius");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemBackgroundColor", {
        get: function () {
            return this.a6;
        },
        set: function (a) {
            this.a6 = a;
            this.e("ItemBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemDisabledBackgroundColor", {
        get: function () {
            return this.a8;
        },
        set: function (a) {
            this.a8 = a;
            this.e("ItemDisabledBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemTextColor", {
        get: function () {
            return this.bd;
        },
        set: function (a) {
            this.bd = a;
            this.e("ItemTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemDisabledTextColor", {
        get: function () {
            return this.ba;
        },
        set: function (a) {
            this.ba = a;
            this.e("ItemDisabledTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemHoverBackgroundColor", {
        get: function () {
            return this.bb;
        },
        set: function (a) {
            this.bb = a;
            this.e("ItemHoverBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemBorderColor", {
        get: function () {
            return this.a7;
        },
        set: function (a) {
            this.a7 = a;
            this.e("ItemBorderColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemDisabledBorderColor", {
        get: function () {
            return this.a9;
        },
        set: function (a) {
            this.a9 = a;
            this.e("ItemDisabledBorderColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemBorderWidth", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.e("ItemBorderWidth");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "itemHoverTextColor", {
        get: function () {
            return this.bc;
        },
        set: function (a) {
            this.bc = a;
            this.e("ItemHoverTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "selectedItemBackgroundColor", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            this.br = a;
            this.e("SelectedItemBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "selectedItemTextColor", {
        get: function () {
            return this.bu;
        },
        set: function (a) {
            this.bu = a;
            this.e("SelectedItemTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "selectedItemHoverBackgroundColor", {
        get: function () {
            return this.bs;
        },
        set: function (a) {
            this.bs = a;
            this.e("SelectedItemHoverBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "selectedItemHoverTextColor", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            this.bt = a;
            this.e("SelectedItemHoverTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemCornerRadius", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.e("OutlinedItemCornerRadius");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemBackgroundColor", {
        get: function () {
            return this.bf;
        },
        set: function (a) {
            this.bf = a;
            this.e("OutlinedItemBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemDisabledBackgroundColor", {
        get: function () {
            return this.bh;
        },
        set: function (a) {
            this.bh = a;
            this.e("OutlinedItemDisabledBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemTextColor", {
        get: function () {
            return this.bm;
        },
        set: function (a) {
            this.bm = a;
            this.e("OutlinedItemTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemDisabledTextColor", {
        get: function () {
            return this.bj;
        },
        set: function (a) {
            this.bj = a;
            this.e("OutlinedItemDisabledTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemHoverBackgroundColor", {
        get: function () {
            return this.bk;
        },
        set: function (a) {
            this.bk = a;
            this.e("OutlinedItemHoverBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemBorderColor", {
        get: function () {
            return this.bg;
        },
        set: function (a) {
            this.bg = a;
            this.e("OutlinedItemBorderColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemDisabledBorderColor", {
        get: function () {
            return this.bi;
        },
        set: function (a) {
            this.bi = a;
            this.e("OutlinedItemDisabledBorderColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemBorderWidth", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.e("OutlinedItemBorderWidth");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedItemHoverTextColor", {
        get: function () {
            return this.bl;
        },
        set: function (a) {
            this.bl = a;
            this.e("OutlinedItemHoverTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedSelectedItemBackgroundColor", {
        get: function () {
            return this.bn;
        },
        set: function (a) {
            this.bn = a;
            this.e("OutlinedSelectedItemBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedSelectedItemTextColor", {
        get: function () {
            return this.bq;
        },
        set: function (a) {
            this.bq = a;
            this.e("OutlinedSelectedItemTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedSelectedItemHoverBackgroundColor", {
        get: function () {
            return this.bo;
        },
        set: function (a) {
            this.bo = a;
            this.e("OutlinedSelectedItemHoverBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "outlinedSelectedItemHoverTextColor", {
        get: function () {
            return this.bp;
        },
        set: function (a) {
            this.bp = a;
            this.e("OutlinedSelectedItemHoverTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemCornerRadius", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.e("FlatItemCornerRadius");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemBackgroundColor", {
        get: function () {
            return this.at;
        },
        set: function (a) {
            this.at = a;
            this.e("FlatItemBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemDisabledBackgroundColor", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.e("FlatItemDisabledBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemTextColor", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.e("FlatItemTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemDisabledTextColor", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.e("FlatItemDisabledTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemHoverBackgroundColor", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.e("FlatItemHoverBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemBorderColor", {
        get: function () {
            return this.au;
        },
        set: function (a) {
            this.au = a;
            this.e("FlatItemBorderColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemDisabledBorderColor", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.e("FlatItemDisabledBorderColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemBorderWidth", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.e("FlatItemBorderWidth");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatItemHoverTextColor", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.e("FlatItemHoverTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatSelectedItemBackgroundColor", {
        get: function () {
            return this.a1;
        },
        set: function (a) {
            this.a1 = a;
            this.e("FlatSelectedItemBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatSelectedItemTextColor", {
        get: function () {
            return this.a4;
        },
        set: function (a) {
            this.a4 = a;
            this.e("FlatSelectedItemTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatSelectedItemHoverBackgroundColor", {
        get: function () {
            return this.a2;
        },
        set: function (a) {
            this.a2 = a;
            this.e("FlatSelectedItemHoverBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "flatSelectedItemHoverTextColor", {
        get: function () {
            return this.a3;
        },
        set: function (a) {
            this.a3 = a;
            this.e("FlatSelectedItemHoverTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "orientation", {
        get: function () {
            return this.be;
        },
        set: function (a) {
            this.be = a;
            this.e("Orientation");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "id", {
        get: function () {
            return this.a5;
        },
        set: function (a) {
            this.a5 = a;
            this.e("Id");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "disabled", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("Disabled");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "isMultiSelect", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("IsMultiSelect");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "buttons", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("Buttons");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonGroupDescription.prototype, "selectionChangedRef", {
        get: function () {
            return this.bv;
        },
        set: function (a) {
            this.bv = a;
            this.e("SelectionChangedRef");
        },
        enumerable: true,
        configurable: true
    });
    ButtonGroupDescription.$t = markType(ButtonGroupDescription, 'ButtonGroupDescription', Description.$);
    return ButtonGroupDescription;
}(Description));
export { ButtonGroupDescription };
