// igniteui-react-core/esm5/lib/CalloutSeriesSelectingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var CalloutSeriesSelectingEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(CalloutSeriesSelectingEventArgsDescription, _super);
    function CalloutSeriesSelectingEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.j = null;
        _this.k = null;
        _this.h = null;
        _this.f = null;
        _this.i = null;
        return _this;
    }
    CalloutSeriesSelectingEventArgsDescription.prototype.get_type = function () {
        return "CalloutSeriesSelectingEventArgs";
    };
    Object.defineProperty(CalloutSeriesSelectingEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutSeriesSelectingEventArgsDescription.prototype, "xValueRef", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("XValueRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutSeriesSelectingEventArgsDescription.prototype, "yValueRef", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("YValueRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutSeriesSelectingEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("ItemRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutSeriesSelectingEventArgsDescription.prototype, "series", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("Series");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CalloutSeriesSelectingEventArgsDescription.prototype, "seriesName", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("SeriesName");
        },
        enumerable: true,
        configurable: true
    });
    CalloutSeriesSelectingEventArgsDescription.$t = markType(CalloutSeriesSelectingEventArgsDescription, 'CalloutSeriesSelectingEventArgsDescription', Description.$);
    return CalloutSeriesSelectingEventArgsDescription;
}(Description));
export { CalloutSeriesSelectingEventArgsDescription };
