// igniteui-react-core/esm5/lib/ColumnPinningDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ColumnPinningDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(ColumnPinningDescription, _super);
    function ColumnPinningDescription() {
        var _this = _super.call(this) || this;
        _this.l = null;
        _this.m = null;
        _this.o = null;
        _this.f = 0;
        _this.p = null;
        _this.q = null;
        _this.n = null;
        _this.k = null;
        _this.i = null;
        _this.j = null;
        _this.h = null;
        return _this;
    }
    ColumnPinningDescription.prototype.get_type = function () {
        return "ColumnPinning";
    };
    Object.defineProperty(ColumnPinningDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnPinningDescription.prototype, "targetGridRef", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("TargetGridRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnPinningDescription.prototype, "title", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.e("Title");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnPinningDescription.prototype, "titleFontFamily", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.e("TitleFontFamily");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnPinningDescription.prototype, "titleFontSize", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("TitleFontSize");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnPinningDescription.prototype, "titleFontStyle", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.e("TitleFontStyle");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnPinningDescription.prototype, "titleFontWeight", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.e("TitleFontWeight");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnPinningDescription.prototype, "titleColor", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.e("TitleColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnPinningDescription.prototype, "filterPlaceholderText", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("FilterPlaceholderText");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnPinningDescription.prototype, "baseTheme", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("BaseTheme");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnPinningDescription.prototype, "density", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("Density");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnPinningDescription.prototype, "backgroundColor", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("BackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    ColumnPinningDescription.$t = markType(ColumnPinningDescription, 'ColumnPinningDescription', Description.$);
    return ColumnPinningDescription;
}(Description));
export { ColumnPinningDescription };
