// igniteui-react-core/esm5/lib/DataSourceSpecialRowDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataSourceSpecialRowDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(DataSourceSpecialRowDescription, _super);
    function DataSourceSpecialRowDescription() {
        var _this = _super.call(this) || this;
        _this.n = null;
        _this.k = null;
        _this.i = 0;
        _this.f = null;
        return _this;
    }
    DataSourceSpecialRowDescription.prototype.get_type = function () {
        return "DataSourceSpecialRow";
    };
    Object.defineProperty(DataSourceSpecialRowDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataSourceSpecialRowDescription.prototype, "rowType", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.e("RowType");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataSourceSpecialRowDescription.prototype, "targetRow", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("TargetRow");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataSourceSpecialRowDescription.prototype, "level", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("Level");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataSourceSpecialRowDescription.prototype, "summaryResults", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("SummaryResults");
        },
        enumerable: true,
        configurable: true
    });
    DataSourceSpecialRowDescription.$t = markType(DataSourceSpecialRowDescription, 'DataSourceSpecialRowDescription', Description.$);
    DataSourceSpecialRowDescription.__marshalByValue = true;
    DataSourceSpecialRowDescription.__marshalByValueAlias = "DataSourceSpecialRow";
    return DataSourceSpecialRowDescription;
}(Description));
export { DataSourceSpecialRowDescription };
