// igniteui-react-core/esm5/lib/DataGridSortIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var DataGridSortIndicatorDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(DataGridSortIndicatorDescription, _super);
    function DataGridSortIndicatorDescription() {
        var _this = _super.call(this) || this;
        _this.n = null;
        _this.l = null;
        _this.f = false;
        _this.i = 0;
        _this.h = 0;
        _this.o = null;
        _this.m = null;
        return _this;
    }
    DataGridSortIndicatorDescription.prototype.get_type = function () {
        return "DataGridSortIndicator";
    };
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "sortDirection", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.e("SortDirection");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "iconColor", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("IconColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "isAnimationEnabled", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("IsAnimationEnabled");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "pixelScalingRatio", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("PixelScalingRatio");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "actualPixelScalingRatio", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("ActualPixelScalingRatio");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "sortIndicatorStyle", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.e("SortIndicatorStyle");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataGridSortIndicatorDescription.prototype, "renderCompletedRef", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.e("RenderCompletedRef");
        },
        enumerable: true,
        configurable: true
    });
    DataGridSortIndicatorDescription.$t = markType(DataGridSortIndicatorDescription, 'DataGridSortIndicatorDescription', Description.$);
    return DataGridSortIndicatorDescription;
}(Description));
export { DataGridSortIndicatorDescription };
