// igniteui-react-core/esm5/lib/DataContext.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Base, INotifyPropertyChanged_$type, PropertyChangedEventArgs, markType } from "./type";
import { ILegendContext_$type } from "./ILegendContext";
/**
 * @hidden
 */
var DataContext = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(DataContext, _super);
    function DataContext() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.f = null;
        _this.c = null;
        _this.i = null;
        _this.k = null;
        _this.d = null;
        _this.j = null;
        _this.a = 0;
        _this.e = null;
        _this.propertyChanged = null;
        return _this;
    }
    Object.defineProperty(DataContext.prototype, "series", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "item", {
        get: function () {
            return this.c;
        },
        set: function (a) {
            if (a != this.c) {
                this.c = a;
                this.h("Item");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "actualItemBrush", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            if (a != this.i) {
                this.i = a;
                this.h("ActualItemBrush");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "outline", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            if (a != this.k) {
                this.k = a;
                this.h("Outline");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "itemLabel", {
        get: function () {
            return this.d;
        },
        set: function (a) {
            if (a != this.d) {
                this.d = a;
                this.h("ItemLabel");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "itemBrush", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            if (a != this.j) {
                this.j = a;
                this.h("ItemBrush");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "thickness", {
        get: function () {
            return this.a;
        },
        set: function (a) {
            if (a != this.a) {
                this.a = a;
                this.h("Thickness");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataContext.prototype, "legendLabel", {
        get: function () {
            return this.e;
        },
        set: function (a) {
            if (a != this.e) {
                this.e = a;
                this.h("LegendLabel");
            }
        },
        enumerable: true,
        configurable: true
    });
    DataContext.prototype.h = function (a) {
        if (this.propertyChanged != null) {
            this.propertyChanged(this, new PropertyChangedEventArgs(a));
        }
    };
    DataContext.prototype.flatten = function () {
        var series_ = this.series;
        var item_ = this.item;
        var actualItemBrush_ = this.actualItemBrush;
        var outline_ = this.outline;
        var itemLabel_ = this.itemLabel;
        var itemBrush_ = this.itemBrush;
        var thickness_ = this.thickness;
        return { item: item_, itemBrush: actualItemBrush_, outline: outline_, itemLabel: itemLabel_, thickness: thickness_ };
    };
    DataContext.$t = markType(DataContext, 'DataContext', Base.$, [ILegendContext_$type, INotifyPropertyChanged_$type]);
    return DataContext;
}(Base));
export { DataContext };
