// igniteui-react-core/esm5/lib/ExpansionPanelDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var ExpansionPanelDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(ExpansionPanelDescription, _super);
    function ExpansionPanelDescription() {
        var _this = _super.call(this) || this;
        _this.r = null;
        _this.u = null;
        _this.m = null;
        _this.s = null;
        _this.t = null;
        _this.v = null;
        _this.y = null;
        _this.n = null;
        _this.w = null;
        _this.x = null;
        _this.i = 0;
        _this.h = 0;
        _this.f = false;
        _this.z = null;
        _this.o = null;
        _this.aa = null;
        _this.ab = null;
        _this.q = null;
        _this.p = null;
        _this.l = null;
        _this.ac = null;
        _this.ad = null;
        return _this;
    }
    ExpansionPanelDescription.prototype.get_type = function () {
        return "ExpansionPanel";
    };
    Object.defineProperty(ExpansionPanelDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "caption", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.e("Caption");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "captionTextColor", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.e("CaptionTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "actualCaptionTextColor", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.e("ActualCaptionTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "captionCollapsedTextColor", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.e("CaptionCollapsedTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "captionExpandedTextColor", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.e("CaptionExpandedTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "description", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.e("Description");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "descriptionTextColor", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.e("DescriptionTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "actualDescriptionTextColor", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.e("ActualDescriptionTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "descriptionCollapsedTextColor", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.e("DescriptionCollapsedTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "descriptionExpandedTextColor", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.e("DescriptionExpandedTextColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "elevation", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("Elevation");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "actualElevation", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("ActualElevation");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "expanded", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("Expanded");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "headerBackgroundColor", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            this.z = a;
            this.e("HeaderBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "actualHeaderBackgroundColor", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.e("ActualHeaderBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "headerCollapsedBackgroundColor", {
        get: function () {
            return this.aa;
        },
        set: function (a) {
            this.aa = a;
            this.e("HeaderCollapsedBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "headerExpandedBackgroundColor", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            this.ab = a;
            this.e("HeaderExpandedBackgroundColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "actualUmbraShadowColor", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.e("ActualUmbraShadowColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "actualPenumbraShadowColor", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.e("ActualPenumbraShadowColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "actualAmbientShadowColor", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("ActualAmbientShadowColor");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "onCollapsedRef", {
        get: function () {
            return this.ac;
        },
        set: function (a) {
            this.ac = a;
            this.e("OnCollapsedRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExpansionPanelDescription.prototype, "onExpandedRef", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            this.ad = a;
            this.e("OnExpandedRef");
        },
        enumerable: true,
        configurable: true
    });
    ExpansionPanelDescription.$t = markType(ExpansionPanelDescription, 'ExpansionPanelDescription', Description.$);
    return ExpansionPanelDescription;
}(Description));
export { ExpansionPanelDescription };
