// igniteui-react-core/esm5/lib/GridCellValueChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var GridCellValueChangedEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(GridCellValueChangedEventArgsDescription, _super);
    function GridCellValueChangedEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.h = 0;
        _this.f = null;
        _this.l = null;
        _this.k = null;
        _this.j = null;
        _this.m = null;
        return _this;
    }
    GridCellValueChangedEventArgsDescription.prototype.get_type = function () {
        return "GridCellValueChangedEventArgs";
    };
    Object.defineProperty(GridCellValueChangedEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridCellValueChangedEventArgsDescription.prototype, "transactionID", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("TransactionID");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridCellValueChangedEventArgsDescription.prototype, "column", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("Column");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridCellValueChangedEventArgsDescription.prototype, "oldValueRef", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("OldValueRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridCellValueChangedEventArgsDescription.prototype, "newValueRef", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("NewValueRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridCellValueChangedEventArgsDescription.prototype, "errorMessage", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("ErrorMessage");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridCellValueChangedEventArgsDescription.prototype, "pendingState", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.e("PendingState");
        },
        enumerable: true,
        configurable: true
    });
    GridCellValueChangedEventArgsDescription.$t = markType(GridCellValueChangedEventArgsDescription, 'GridCellValueChangedEventArgsDescription', Description.$);
    return GridCellValueChangedEventArgsDescription;
}(Description));
export { GridCellValueChangedEventArgsDescription };
