// igniteui-react-core/esm5/lib/GridCellEditEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var GridCellEditEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(GridCellEditEventArgsDescription, _super);
    function GridCellEditEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.f = null;
        _this.l = null;
        _this.k = null;
        _this.j = null;
        _this.h = false;
        return _this;
    }
    GridCellEditEventArgsDescription.prototype.get_type = function () {
        return "GridCellEditEventArgs";
    };
    Object.defineProperty(GridCellEditEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridCellEditEventArgsDescription.prototype, "column", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("Column");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridCellEditEventArgsDescription.prototype, "oldValueRef", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("OldValueRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridCellEditEventArgsDescription.prototype, "newValueRef", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("NewValueRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridCellEditEventArgsDescription.prototype, "errorMessage", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("ErrorMessage");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridCellEditEventArgsDescription.prototype, "cancel", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("Cancel");
        },
        enumerable: true,
        configurable: true
    });
    GridCellEditEventArgsDescription.$t = markType(GridCellEditEventArgsDescription, 'GridCellEditEventArgsDescription', Description.$);
    return GridCellEditEventArgsDescription;
}(Description));
export { GridCellEditEventArgsDescription };
