// igniteui-react-core/esm5/lib/GridFilterDialogViewModelRowDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var GridFilterDialogViewModelRowDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(GridFilterDialogViewModelRowDescription, _super);
    function GridFilterDialogViewModelRowDescription() {
        var _this = _super.call(this) || this;
        _this.g = null;
        _this.j = null;
        _this.f = null;
        _this.k = null;
        _this.l = null;
        return _this;
    }
    GridFilterDialogViewModelRowDescription.prototype.get_type = function () {
        return "GridFilterDialogViewModelRow";
    };
    Object.defineProperty(GridFilterDialogViewModelRowDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridFilterDialogViewModelRowDescription.prototype, "operators", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("Operators");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridFilterDialogViewModelRowDescription.prototype, "currentOperator", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("CurrentOperator");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridFilterDialogViewModelRowDescription.prototype, "operandNumber", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("OperandNumber");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridFilterDialogViewModelRowDescription.prototype, "operand1Ref", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("Operand1Ref");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GridFilterDialogViewModelRowDescription.prototype, "operand2Ref", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("Operand2Ref");
        },
        enumerable: true,
        configurable: true
    });
    GridFilterDialogViewModelRowDescription.$t = markType(GridFilterDialogViewModelRowDescription, 'GridFilterDialogViewModelRowDescription', Description.$);
    return GridFilterDialogViewModelRowDescription;
}(Description));
export { GridFilterDialogViewModelRowDescription };
