// igniteui-react-core/esm5/lib/PrefixDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { InputGroupItemDescription } from "./InputGroupItemDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var PrefixDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(PrefixDescription, _super);
    function PrefixDescription() {
        var _this = _super.call(this) || this;
        _this.n = null;
        _this.m = null;
        _this.j = false;
        _this.i = false;
        return _this;
    }
    PrefixDescription.prototype.get_type = function () {
        return "Prefix";
    };
    Object.defineProperty(PrefixDescription.prototype, "id", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.e("Id");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PrefixDescription.prototype, "ariaLabel", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.e("AriaLabel");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PrefixDescription.prototype, "isHover", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("IsHover");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PrefixDescription.prototype, "disabled", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("Disabled");
        },
        enumerable: true,
        configurable: true
    });
    PrefixDescription.$t = markType(PrefixDescription, 'PrefixDescription', InputGroupItemDescription.$);
    return PrefixDescription;
}(InputGroupItemDescription));
export { PrefixDescription };
