// igniteui-react-core/esm5/lib/StyleShapeEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var StyleShapeEventArgsDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(StyleShapeEventArgsDescription, _super);
    function StyleShapeEventArgsDescription() {
        var _this = _super.call(this) || this;
        _this.j = null;
        _this.k = null;
        _this.l = null;
        _this.g = 0;
        _this.f = 0;
        return _this;
    }
    StyleShapeEventArgsDescription.prototype.get_type = function () {
        return "StyleShapeEventArgs";
    };
    Object.defineProperty(StyleShapeEventArgsDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StyleShapeEventArgsDescription.prototype, "itemRef", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("ItemRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StyleShapeEventArgsDescription.prototype, "shapeFill", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.e("ShapeFill");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StyleShapeEventArgsDescription.prototype, "shapeStroke", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.e("ShapeStroke");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StyleShapeEventArgsDescription.prototype, "shapeStrokeThickness", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("ShapeStrokeThickness");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StyleShapeEventArgsDescription.prototype, "shapeOpacity", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("ShapeOpacity");
        },
        enumerable: true,
        configurable: true
    });
    StyleShapeEventArgsDescription.$t = markType(StyleShapeEventArgsDescription, 'StyleShapeEventArgsDescription', Description.$);
    return StyleShapeEventArgsDescription;
}(Description));
export { StyleShapeEventArgsDescription };
