// igniteui-react-core/esm5/lib/SummaryDataDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var SummaryDataDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(SummaryDataDescription, _super);
    function SummaryDataDescription() {
        var _this = _super.call(this) || this;
        _this.h = null;
        _this.j = null;
        _this.i = null;
        _this.f = null;
        _this.g = null;
        return _this;
    }
    SummaryDataDescription.prototype.get_type = function () {
        return "SummaryData";
    };
    Object.defineProperty(SummaryDataDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SummaryDataDescription.prototype, "summaryName", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("SummaryName");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SummaryDataDescription.prototype, "summaryValueRef", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.e("SummaryValueRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SummaryDataDescription.prototype, "summaryOperand", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("SummaryOperand");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SummaryDataDescription.prototype, "formattedText", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("FormattedText");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SummaryDataDescription.prototype, "formattedValue", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("FormattedValue");
        },
        enumerable: true,
        configurable: true
    });
    SummaryDataDescription.$t = markType(SummaryDataDescription, 'SummaryDataDescription', Description.$);
    return SummaryDataDescription;
}(Description));
export { SummaryDataDescription };
