// igniteui-react-core/esm5/lib/TransactionStateDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import * as tslib_1 from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var TransactionStateDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    tslib_1.__extends(TransactionStateDescription, _super);
    function TransactionStateDescription() {
        var _this = _super.call(this) || this;
        _this.f = null;
        _this.g = null;
        _this.h = null;
        _this.i = null;
        return _this;
    }
    TransactionStateDescription.prototype.get_type = function () {
        return "TransactionState";
    };
    Object.defineProperty(TransactionStateDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TransactionStateDescription.prototype, "idRef", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.e("IdRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TransactionStateDescription.prototype, "transactionType", {
        get: function () {
            return this.g;
        },
        set: function (a) {
            this.g = a;
            this.e("TransactionType");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TransactionStateDescription.prototype, "valueRef", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.e("ValueRef");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TransactionStateDescription.prototype, "versionRef", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.e("VersionRef");
        },
        enumerable: true,
        configurable: true
    });
    TransactionStateDescription.$t = markType(TransactionStateDescription, 'TransactionStateDescription', Description.$);
    return TransactionStateDescription;
}(Description));
export { TransactionStateDescription };
