// igniteui-react-core/esm2015/lib/ButtonDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ButtonDescription = /*@__PURE__*/ (() => {
    class ButtonDescription extends Description {
        constructor() {
            super();
            this.au = 0;
            this.at = 0;
            this.el = null;
            this.eg = null;
            this.ed = null;
            this.dx = null;
            this.dw = null;
            this.d9 = null;
            this.d6 = null;
            this.du = null;
            this.ca = 0;
            this.b8 = 0;
            this.b9 = 0;
            this.d7 = null;
            this.v = 0;
            this.u = 0;
            this.w = 0;
            this.x = 0;
            this.dv = null;
            this.dy = null;
            this.dz = null;
            this.d0 = null;
            this.b7 = 0;
            this.d1 = null;
            this.d8 = null;
            this.d5 = null;
            this.d2 = null;
            this.d3 = null;
            this.d4 = null;
            this.t = 0;
            this.f7 = null;
            this.f9 = null;
            this.ga = null;
            this.gb = null;
            this.cu = 0;
            this.ge = null;
            this.f8 = null;
            this.cx = 0;
            this.cw = 0;
            this.cv = 0;
            this.a0 = 0;
            this.a2 = 0;
            this.a1 = 0;
            this.a3 = 0;
            this.a4 = 0;
            this.gg = null;
            this.ey = null;
            this.e0 = null;
            this.e1 = null;
            this.e2 = null;
            this.cg = 0;
            this.e5 = null;
            this.ez = null;
            this.cj = 0;
            this.ci = 0;
            this.ch = 0;
            this.e7 = null;
            this.e8 = null;
            this.e6 = null;
            this.e3 = null;
            this.e4 = null;
            this.ai = 0;
            this.ak = 0;
            this.aj = 0;
            this.al = 0;
            this.am = 0;
            this.ec = null;
            this.eh = null;
            this.ei = null;
            this.ej = null;
            this.cb = 0;
            this.em = null;
            this.fg = null;
            this.ee = null;
            this.cy = 0;
            this.cl = 0;
            this.ck = 0;
            this.fd = null;
            this.an = 0;
            this.fe = null;
            this.ff = null;
            this.gj = null;
            this.fh = null;
            this.gi = null;
            this.gh = null;
            this.gf = null;
            this.gc = null;
            this.gd = null;
            this.fb = null;
            this.fc = null;
            this.y = 0;
            this.aa = 0;
            this.z = 0;
            this.ab = 0;
            this.ac = 0;
            this.fw = null;
            this.fy = null;
            this.fz = null;
            this.f0 = null;
            this.cq = 0;
            this.f3 = null;
            this.fx = null;
            this.ct = 0;
            this.cs = 0;
            this.cr = 0;
            this.av = 0;
            this.ax = 0;
            this.aw = 0;
            this.ay = 0;
            this.az = 0;
            this.f5 = null;
            this.f6 = null;
            this.f4 = null;
            this.f1 = null;
            this.f2 = null;
            this.en = null;
            this.ep = null;
            this.eq = null;
            this.er = null;
            this.cc = 0;
            this.eu = null;
            this.eo = null;
            this.cf = 0;
            this.ce = 0;
            this.cd = 0;
            this.ad = 0;
            this.af = 0;
            this.ae = 0;
            this.ag = 0;
            this.ah = 0;
            this.ew = null;
            this.ex = null;
            this.ev = null;
            this.es = null;
            this.et = null;
            this.fi = null;
            this.fk = null;
            this.fl = null;
            this.fm = null;
            this.cm = 0;
            this.fp = null;
            this.fj = null;
            this.cp = 0;
            this.co = 0;
            this.cn = 0;
            this.ao = 0;
            this.aq = 0;
            this.ap = 0;
            this.ar = 0;
            this.as = 0;
            this.fr = null;
            this.fs = null;
            this.fq = null;
            this.fn = null;
            this.fo = null;
            this.fu = null;
            this.ft = null;
            this.ek = null;
            this.e9 = null;
            this.fa = null;
            this.ea = null;
            this.gk = null;
            this.fv = null;
            this.cz = 0;
            this.j = false;
            this.k = false;
            this.eb = null;
            this.h = false;
            this.f = false;
            this.l = false;
            this.g = false;
            this.i = false;
            this.ef = null;
        }
        get_type() {
            return "Button";
        }
        get type() {
            return this.get_type();
        }
        get minWidth() {
            return this.au;
        }
        set minWidth(a) {
            this.au = a;
            this.e("MinWidth");
        }
        get minHeight() {
            return this.at;
        }
        set minHeight(a) {
            this.at = a;
            this.e("MinHeight");
        }
        get displayType() {
            return this.el;
        }
        set displayType(a) {
            this.el = a;
            this.e("DisplayType");
        }
        get density() {
            return this.eg;
        }
        set density(a) {
            this.eg = a;
            this.e("Density");
        }
        get baseTheme() {
            return this.ed;
        }
        set baseTheme(a) {
            this.ed = a;
            this.e("BaseTheme");
        }
        get actualDensity() {
            return this.dx;
        }
        set actualDensity(a) {
            this.dx = a;
            this.e("ActualDensity");
        }
        get actualBorderColor() {
            return this.dw;
        }
        set actualBorderColor(a) {
            this.dw = a;
            this.e("ActualBorderColor");
        }
        get actualUmbraShadowColor() {
            return this.d9;
        }
        set actualUmbraShadowColor(a) {
            this.d9 = a;
            this.e("ActualUmbraShadowColor");
        }
        get actualPenumbraShadowColor() {
            return this.d6;
        }
        set actualPenumbraShadowColor(a) {
            this.d6 = a;
            this.e("ActualPenumbraShadowColor");
        }
        get actualAmbientShadowColor() {
            return this.du;
        }
        set actualAmbientShadowColor(a) {
            this.du = a;
            this.e("ActualAmbientShadowColor");
        }
        get actualRestingElevation() {
            return this.ca;
        }
        set actualRestingElevation(a) {
            this.ca = a;
            this.e("ActualRestingElevation");
        }
        get actualFocusElevation() {
            return this.b8;
        }
        set actualFocusElevation(a) {
            this.b8 = a;
            this.e("ActualFocusElevation");
        }
        get actualHoverElevation() {
            return this.b9;
        }
        set actualHoverElevation(a) {
            this.b9 = a;
            this.e("ActualHoverElevation");
        }
        get actualRippleColor() {
            return this.d7;
        }
        set actualRippleColor(a) {
            this.d7 = a;
            this.e("ActualRippleColor");
        }
        get actualCornerRadiusBottomRight() {
            return this.v;
        }
        set actualCornerRadiusBottomRight(a) {
            this.v = a;
            this.e("ActualCornerRadiusBottomRight");
        }
        get actualCornerRadiusBottomLeft() {
            return this.u;
        }
        set actualCornerRadiusBottomLeft(a) {
            this.u = a;
            this.e("ActualCornerRadiusBottomLeft");
        }
        get actualCornerRadiusTopLeft() {
            return this.w;
        }
        set actualCornerRadiusTopLeft(a) {
            this.w = a;
            this.e("ActualCornerRadiusTopLeft");
        }
        get actualCornerRadiusTopRight() {
            return this.x;
        }
        set actualCornerRadiusTopRight(a) {
            this.x = a;
            this.e("ActualCornerRadiusTopRight");
        }
        get actualBackgroundColor() {
            return this.dv;
        }
        set actualBackgroundColor(a) {
            this.dv = a;
            this.e("ActualBackgroundColor");
        }
        get actualDisabledBackgroundColor() {
            return this.dy;
        }
        set actualDisabledBackgroundColor(a) {
            this.dy = a;
            this.e("ActualDisabledBackgroundColor");
        }
        get actualDisabledBorderColor() {
            return this.dz;
        }
        set actualDisabledBorderColor(a) {
            this.dz = a;
            this.e("ActualDisabledBorderColor");
        }
        get actualDisabledTextColor() {
            return this.d0;
        }
        set actualDisabledTextColor(a) {
            this.d0 = a;
            this.e("ActualDisabledTextColor");
        }
        get actualDisabledElevation() {
            return this.b7;
        }
        set actualDisabledElevation(a) {
            this.b7 = a;
            this.e("ActualDisabledElevation");
        }
        get actualElevationMode() {
            return this.d1;
        }
        set actualElevationMode(a) {
            this.d1 = a;
            this.e("ActualElevationMode");
        }
        get actualTextColor() {
            return this.d8;
        }
        set actualTextColor(a) {
            this.d8 = a;
            this.e("ActualTextColor");
        }
        get actualHoverTextColor() {
            return this.d5;
        }
        set actualHoverTextColor(a) {
            this.d5 = a;
            this.e("ActualHoverTextColor");
        }
        get actualFocusBackgroundColor() {
            return this.d2;
        }
        set actualFocusBackgroundColor(a) {
            this.d2 = a;
            this.e("ActualFocusBackgroundColor");
        }
        get actualFocusTextColor() {
            return this.d3;
        }
        set actualFocusTextColor(a) {
            this.d3 = a;
            this.e("ActualFocusTextColor");
        }
        get actualHoverBackgroundColor() {
            return this.d4;
        }
        set actualHoverBackgroundColor(a) {
            this.d4 = a;
            this.e("ActualHoverBackgroundColor");
        }
        get actualBorderWidth() {
            return this.t;
        }
        set actualBorderWidth(a) {
            this.t = a;
            this.e("ActualBorderWidth");
        }
        get raisedBackgroundColor() {
            return this.f7;
        }
        set raisedBackgroundColor(a) {
            this.f7 = a;
            this.e("RaisedBackgroundColor");
        }
        get raisedDisabledBackgroundColor() {
            return this.f9;
        }
        set raisedDisabledBackgroundColor(a) {
            this.f9 = a;
            this.e("RaisedDisabledBackgroundColor");
        }
        get raisedDisabledBorderColor() {
            return this.ga;
        }
        set raisedDisabledBorderColor(a) {
            this.ga = a;
            this.e("RaisedDisabledBorderColor");
        }
        get raisedDisabledTextColor() {
            return this.gb;
        }
        set raisedDisabledTextColor(a) {
            this.gb = a;
            this.e("RaisedDisabledTextColor");
        }
        get raisedDisabledElevation() {
            return this.cu;
        }
        set raisedDisabledElevation(a) {
            this.cu = a;
            this.e("RaisedDisabledElevation");
        }
        get raisedHoverBackgroundColor() {
            return this.ge;
        }
        set raisedHoverBackgroundColor(a) {
            this.ge = a;
            this.e("RaisedHoverBackgroundColor");
        }
        get raisedBorderColor() {
            return this.f8;
        }
        set raisedBorderColor(a) {
            this.f8 = a;
            this.e("RaisedBorderColor");
        }
        get raisedRestingElevation() {
            return this.cx;
        }
        set raisedRestingElevation(a) {
            this.cx = a;
            this.e("RaisedRestingElevation");
        }
        get raisedHoverElevation() {
            return this.cw;
        }
        set raisedHoverElevation(a) {
            this.cw = a;
            this.e("RaisedHoverElevation");
        }
        get raisedFocusElevation() {
            return this.cv;
        }
        set raisedFocusElevation(a) {
            this.cv = a;
            this.e("RaisedFocusElevation");
        }
        get raisedBorderWidth() {
            return this.a0;
        }
        set raisedBorderWidth(a) {
            this.a0 = a;
            this.e("RaisedBorderWidth");
        }
        get raisedCornerRadiusBottomRight() {
            return this.a2;
        }
        set raisedCornerRadiusBottomRight(a) {
            this.a2 = a;
            this.e("RaisedCornerRadiusBottomRight");
        }
        get raisedCornerRadiusBottomLeft() {
            return this.a1;
        }
        set raisedCornerRadiusBottomLeft(a) {
            this.a1 = a;
            this.e("RaisedCornerRadiusBottomLeft");
        }
        get raisedCornerRadiusTopLeft() {
            return this.a3;
        }
        set raisedCornerRadiusTopLeft(a) {
            this.a3 = a;
            this.e("RaisedCornerRadiusTopLeft");
        }
        get raisedCornerRadiusTopRight() {
            return this.a4;
        }
        set raisedCornerRadiusTopRight(a) {
            this.a4 = a;
            this.e("RaisedCornerRadiusTopRight");
        }
        get raisedRippleColor() {
            return this.gg;
        }
        set raisedRippleColor(a) {
            this.gg = a;
            this.e("RaisedRippleColor");
        }
        get flatBackgroundColor() {
            return this.ey;
        }
        set flatBackgroundColor(a) {
            this.ey = a;
            this.e("FlatBackgroundColor");
        }
        get flatDisabledBackgroundColor() {
            return this.e0;
        }
        set flatDisabledBackgroundColor(a) {
            this.e0 = a;
            this.e("FlatDisabledBackgroundColor");
        }
        get flatDisabledBorderColor() {
            return this.e1;
        }
        set flatDisabledBorderColor(a) {
            this.e1 = a;
            this.e("FlatDisabledBorderColor");
        }
        get flatDisabledTextColor() {
            return this.e2;
        }
        set flatDisabledTextColor(a) {
            this.e2 = a;
            this.e("FlatDisabledTextColor");
        }
        get flatDisabledElevation() {
            return this.cg;
        }
        set flatDisabledElevation(a) {
            this.cg = a;
            this.e("FlatDisabledElevation");
        }
        get flatHoverBackgroundColor() {
            return this.e5;
        }
        set flatHoverBackgroundColor(a) {
            this.e5 = a;
            this.e("FlatHoverBackgroundColor");
        }
        get flatBorderColor() {
            return this.ez;
        }
        set flatBorderColor(a) {
            this.ez = a;
            this.e("FlatBorderColor");
        }
        get flatRestingElevation() {
            return this.cj;
        }
        set flatRestingElevation(a) {
            this.cj = a;
            this.e("FlatRestingElevation");
        }
        get flatHoverElevation() {
            return this.ci;
        }
        set flatHoverElevation(a) {
            this.ci = a;
            this.e("FlatHoverElevation");
        }
        get flatFocusElevation() {
            return this.ch;
        }
        set flatFocusElevation(a) {
            this.ch = a;
            this.e("FlatFocusElevation");
        }
        get flatRippleColor() {
            return this.e7;
        }
        set flatRippleColor(a) {
            this.e7 = a;
            this.e("FlatRippleColor");
        }
        get flatTextColor() {
            return this.e8;
        }
        set flatTextColor(a) {
            this.e8 = a;
            this.e("FlatTextColor");
        }
        get flatHoverTextColor() {
            return this.e6;
        }
        set flatHoverTextColor(a) {
            this.e6 = a;
            this.e("FlatHoverTextColor");
        }
        get flatFocusBackgroundColor() {
            return this.e3;
        }
        set flatFocusBackgroundColor(a) {
            this.e3 = a;
            this.e("FlatFocusBackgroundColor");
        }
        get flatFocusTextColor() {
            return this.e4;
        }
        set flatFocusTextColor(a) {
            this.e4 = a;
            this.e("FlatFocusTextColor");
        }
        get flatBorderWidth() {
            return this.ai;
        }
        set flatBorderWidth(a) {
            this.ai = a;
            this.e("FlatBorderWidth");
        }
        get flatCornerRadiusBottomRight() {
            return this.ak;
        }
        set flatCornerRadiusBottomRight(a) {
            this.ak = a;
            this.e("FlatCornerRadiusBottomRight");
        }
        get flatCornerRadiusBottomLeft() {
            return this.aj;
        }
        set flatCornerRadiusBottomLeft(a) {
            this.aj = a;
            this.e("FlatCornerRadiusBottomLeft");
        }
        get flatCornerRadiusTopLeft() {
            return this.al;
        }
        set flatCornerRadiusTopLeft(a) {
            this.al = a;
            this.e("FlatCornerRadiusTopLeft");
        }
        get flatCornerRadiusTopRight() {
            return this.am;
        }
        set flatCornerRadiusTopRight(a) {
            this.am = a;
            this.e("FlatCornerRadiusTopRight");
        }
        get backgroundColor() {
            return this.ec;
        }
        set backgroundColor(a) {
            this.ec = a;
            this.e("BackgroundColor");
        }
        get disabledBackgroundColor() {
            return this.eh;
        }
        set disabledBackgroundColor(a) {
            this.eh = a;
            this.e("DisabledBackgroundColor");
        }
        get disabledBorderColor() {
            return this.ei;
        }
        set disabledBorderColor(a) {
            this.ei = a;
            this.e("DisabledBorderColor");
        }
        get disabledTextColor() {
            return this.ej;
        }
        set disabledTextColor(a) {
            this.ej = a;
            this.e("DisabledTextColor");
        }
        get disabledElevation() {
            return this.cb;
        }
        set disabledElevation(a) {
            this.cb = a;
            this.e("DisabledElevation");
        }
        get elevationMode() {
            return this.em;
        }
        set elevationMode(a) {
            this.em = a;
            this.e("ElevationMode");
        }
        get hoverBackgroundColor() {
            return this.fg;
        }
        set hoverBackgroundColor(a) {
            this.fg = a;
            this.e("HoverBackgroundColor");
        }
        get borderColor() {
            return this.ee;
        }
        set borderColor(a) {
            this.ee = a;
            this.e("BorderColor");
        }
        get restingElevation() {
            return this.cy;
        }
        set restingElevation(a) {
            this.cy = a;
            this.e("RestingElevation");
        }
        get hoverElevation() {
            return this.cl;
        }
        set hoverElevation(a) {
            this.cl = a;
            this.e("HoverElevation");
        }
        get focusElevation() {
            return this.ck;
        }
        set focusElevation(a) {
            this.ck = a;
            this.e("FocusElevation");
        }
        get fontFamily() {
            return this.fd;
        }
        set fontFamily(a) {
            this.fd = a;
            this.e("FontFamily");
        }
        get fontSize() {
            return this.an;
        }
        set fontSize(a) {
            this.an = a;
            this.e("FontSize");
        }
        get fontStyle() {
            return this.fe;
        }
        set fontStyle(a) {
            this.fe = a;
            this.e("FontStyle");
        }
        get fontWeight() {
            return this.ff;
        }
        set fontWeight(a) {
            this.ff = a;
            this.e("FontWeight");
        }
        get textColor() {
            return this.gj;
        }
        set textColor(a) {
            this.gj = a;
            this.e("TextColor");
        }
        get hoverTextColor() {
            return this.fh;
        }
        set hoverTextColor(a) {
            this.fh = a;
            this.e("HoverTextColor");
        }
        get rippleColor() {
            return this.gi;
        }
        set rippleColor(a) {
            this.gi = a;
            this.e("RippleColor");
        }
        get raisedTextColor() {
            return this.gh;
        }
        set raisedTextColor(a) {
            this.gh = a;
            this.e("RaisedTextColor");
        }
        get raisedHoverTextColor() {
            return this.gf;
        }
        set raisedHoverTextColor(a) {
            this.gf = a;
            this.e("RaisedHoverTextColor");
        }
        get raisedFocusBackgroundColor() {
            return this.gc;
        }
        set raisedFocusBackgroundColor(a) {
            this.gc = a;
            this.e("RaisedFocusBackgroundColor");
        }
        get raisedFocusTextColor() {
            return this.gd;
        }
        set raisedFocusTextColor(a) {
            this.gd = a;
            this.e("RaisedFocusTextColor");
        }
        get focusBackgroundColor() {
            return this.fb;
        }
        set focusBackgroundColor(a) {
            this.fb = a;
            this.e("FocusBackgroundColor");
        }
        get focusTextColor() {
            return this.fc;
        }
        set focusTextColor(a) {
            this.fc = a;
            this.e("FocusTextColor");
        }
        get borderWidth() {
            return this.y;
        }
        set borderWidth(a) {
            this.y = a;
            this.e("BorderWidth");
        }
        get cornerRadiusBottomRight() {
            return this.aa;
        }
        set cornerRadiusBottomRight(a) {
            this.aa = a;
            this.e("CornerRadiusBottomRight");
        }
        get cornerRadiusBottomLeft() {
            return this.z;
        }
        set cornerRadiusBottomLeft(a) {
            this.z = a;
            this.e("CornerRadiusBottomLeft");
        }
        get cornerRadiusTopLeft() {
            return this.ab;
        }
        set cornerRadiusTopLeft(a) {
            this.ab = a;
            this.e("CornerRadiusTopLeft");
        }
        get cornerRadiusTopRight() {
            return this.ac;
        }
        set cornerRadiusTopRight(a) {
            this.ac = a;
            this.e("CornerRadiusTopRight");
        }
        get outlinedBackgroundColor() {
            return this.fw;
        }
        set outlinedBackgroundColor(a) {
            this.fw = a;
            this.e("OutlinedBackgroundColor");
        }
        get outlinedDisabledBackgroundColor() {
            return this.fy;
        }
        set outlinedDisabledBackgroundColor(a) {
            this.fy = a;
            this.e("OutlinedDisabledBackgroundColor");
        }
        get outlinedDisabledBorderColor() {
            return this.fz;
        }
        set outlinedDisabledBorderColor(a) {
            this.fz = a;
            this.e("OutlinedDisabledBorderColor");
        }
        get outlinedDisabledTextColor() {
            return this.f0;
        }
        set outlinedDisabledTextColor(a) {
            this.f0 = a;
            this.e("OutlinedDisabledTextColor");
        }
        get outlinedDisabledElevation() {
            return this.cq;
        }
        set outlinedDisabledElevation(a) {
            this.cq = a;
            this.e("OutlinedDisabledElevation");
        }
        get outlinedHoverBackgroundColor() {
            return this.f3;
        }
        set outlinedHoverBackgroundColor(a) {
            this.f3 = a;
            this.e("OutlinedHoverBackgroundColor");
        }
        get outlinedBorderColor() {
            return this.fx;
        }
        set outlinedBorderColor(a) {
            this.fx = a;
            this.e("OutlinedBorderColor");
        }
        get outlinedRestingElevation() {
            return this.ct;
        }
        set outlinedRestingElevation(a) {
            this.ct = a;
            this.e("OutlinedRestingElevation");
        }
        get outlinedHoverElevation() {
            return this.cs;
        }
        set outlinedHoverElevation(a) {
            this.cs = a;
            this.e("OutlinedHoverElevation");
        }
        get outlinedFocusElevation() {
            return this.cr;
        }
        set outlinedFocusElevation(a) {
            this.cr = a;
            this.e("OutlinedFocusElevation");
        }
        get outlinedBorderWidth() {
            return this.av;
        }
        set outlinedBorderWidth(a) {
            this.av = a;
            this.e("OutlinedBorderWidth");
        }
        get outlinedCornerRadiusBottomRight() {
            return this.ax;
        }
        set outlinedCornerRadiusBottomRight(a) {
            this.ax = a;
            this.e("OutlinedCornerRadiusBottomRight");
        }
        get outlinedCornerRadiusBottomLeft() {
            return this.aw;
        }
        set outlinedCornerRadiusBottomLeft(a) {
            this.aw = a;
            this.e("OutlinedCornerRadiusBottomLeft");
        }
        get outlinedCornerRadiusTopLeft() {
            return this.ay;
        }
        set outlinedCornerRadiusTopLeft(a) {
            this.ay = a;
            this.e("OutlinedCornerRadiusTopLeft");
        }
        get outlinedCornerRadiusTopRight() {
            return this.az;
        }
        set outlinedCornerRadiusTopRight(a) {
            this.az = a;
            this.e("OutlinedCornerRadiusTopRight");
        }
        get outlinedRippleColor() {
            return this.f5;
        }
        set outlinedRippleColor(a) {
            this.f5 = a;
            this.e("OutlinedRippleColor");
        }
        get outlinedTextColor() {
            return this.f6;
        }
        set outlinedTextColor(a) {
            this.f6 = a;
            this.e("OutlinedTextColor");
        }
        get outlinedHoverTextColor() {
            return this.f4;
        }
        set outlinedHoverTextColor(a) {
            this.f4 = a;
            this.e("OutlinedHoverTextColor");
        }
        get outlinedFocusBackgroundColor() {
            return this.f1;
        }
        set outlinedFocusBackgroundColor(a) {
            this.f1 = a;
            this.e("OutlinedFocusBackgroundColor");
        }
        get outlinedFocusTextColor() {
            return this.f2;
        }
        set outlinedFocusTextColor(a) {
            this.f2 = a;
            this.e("OutlinedFocusTextColor");
        }
        get fabBackgroundColor() {
            return this.en;
        }
        set fabBackgroundColor(a) {
            this.en = a;
            this.e("FabBackgroundColor");
        }
        get fabDisabledBackgroundColor() {
            return this.ep;
        }
        set fabDisabledBackgroundColor(a) {
            this.ep = a;
            this.e("FabDisabledBackgroundColor");
        }
        get fabDisabledBorderColor() {
            return this.eq;
        }
        set fabDisabledBorderColor(a) {
            this.eq = a;
            this.e("FabDisabledBorderColor");
        }
        get fabDisabledTextColor() {
            return this.er;
        }
        set fabDisabledTextColor(a) {
            this.er = a;
            this.e("FabDisabledTextColor");
        }
        get fabDisabledElevation() {
            return this.cc;
        }
        set fabDisabledElevation(a) {
            this.cc = a;
            this.e("FabDisabledElevation");
        }
        get fabHoverBackgroundColor() {
            return this.eu;
        }
        set fabHoverBackgroundColor(a) {
            this.eu = a;
            this.e("FabHoverBackgroundColor");
        }
        get fabBorderColor() {
            return this.eo;
        }
        set fabBorderColor(a) {
            this.eo = a;
            this.e("FabBorderColor");
        }
        get fabRestingElevation() {
            return this.cf;
        }
        set fabRestingElevation(a) {
            this.cf = a;
            this.e("FabRestingElevation");
        }
        get fabHoverElevation() {
            return this.ce;
        }
        set fabHoverElevation(a) {
            this.ce = a;
            this.e("FabHoverElevation");
        }
        get fabFocusElevation() {
            return this.cd;
        }
        set fabFocusElevation(a) {
            this.cd = a;
            this.e("FabFocusElevation");
        }
        get fabBorderWidth() {
            return this.ad;
        }
        set fabBorderWidth(a) {
            this.ad = a;
            this.e("FabBorderWidth");
        }
        get fabCornerRadiusBottomRight() {
            return this.af;
        }
        set fabCornerRadiusBottomRight(a) {
            this.af = a;
            this.e("FabCornerRadiusBottomRight");
        }
        get fabCornerRadiusBottomLeft() {
            return this.ae;
        }
        set fabCornerRadiusBottomLeft(a) {
            this.ae = a;
            this.e("FabCornerRadiusBottomLeft");
        }
        get fabCornerRadiusTopLeft() {
            return this.ag;
        }
        set fabCornerRadiusTopLeft(a) {
            this.ag = a;
            this.e("FabCornerRadiusTopLeft");
        }
        get fabCornerRadiusTopRight() {
            return this.ah;
        }
        set fabCornerRadiusTopRight(a) {
            this.ah = a;
            this.e("FabCornerRadiusTopRight");
        }
        get fabRippleColor() {
            return this.ew;
        }
        set fabRippleColor(a) {
            this.ew = a;
            this.e("FabRippleColor");
        }
        get fabTextColor() {
            return this.ex;
        }
        set fabTextColor(a) {
            this.ex = a;
            this.e("FabTextColor");
        }
        get fabHoverTextColor() {
            return this.ev;
        }
        set fabHoverTextColor(a) {
            this.ev = a;
            this.e("FabHoverTextColor");
        }
        get fabFocusBackgroundColor() {
            return this.es;
        }
        set fabFocusBackgroundColor(a) {
            this.es = a;
            this.e("FabFocusBackgroundColor");
        }
        get fabFocusTextColor() {
            return this.et;
        }
        set fabFocusTextColor(a) {
            this.et = a;
            this.e("FabFocusTextColor");
        }
        get iconBackgroundColor() {
            return this.fi;
        }
        set iconBackgroundColor(a) {
            this.fi = a;
            this.e("IconBackgroundColor");
        }
        get iconDisabledBackgroundColor() {
            return this.fk;
        }
        set iconDisabledBackgroundColor(a) {
            this.fk = a;
            this.e("IconDisabledBackgroundColor");
        }
        get iconDisabledBorderColor() {
            return this.fl;
        }
        set iconDisabledBorderColor(a) {
            this.fl = a;
            this.e("IconDisabledBorderColor");
        }
        get iconDisabledTextColor() {
            return this.fm;
        }
        set iconDisabledTextColor(a) {
            this.fm = a;
            this.e("IconDisabledTextColor");
        }
        get iconDisabledElevation() {
            return this.cm;
        }
        set iconDisabledElevation(a) {
            this.cm = a;
            this.e("IconDisabledElevation");
        }
        get iconHoverBackgroundColor() {
            return this.fp;
        }
        set iconHoverBackgroundColor(a) {
            this.fp = a;
            this.e("IconHoverBackgroundColor");
        }
        get iconBorderColor() {
            return this.fj;
        }
        set iconBorderColor(a) {
            this.fj = a;
            this.e("IconBorderColor");
        }
        get iconRestingElevation() {
            return this.cp;
        }
        set iconRestingElevation(a) {
            this.cp = a;
            this.e("IconRestingElevation");
        }
        get iconHoverElevation() {
            return this.co;
        }
        set iconHoverElevation(a) {
            this.co = a;
            this.e("IconHoverElevation");
        }
        get iconFocusElevation() {
            return this.cn;
        }
        set iconFocusElevation(a) {
            this.cn = a;
            this.e("IconFocusElevation");
        }
        get iconBorderWidth() {
            return this.ao;
        }
        set iconBorderWidth(a) {
            this.ao = a;
            this.e("IconBorderWidth");
        }
        get iconCornerRadiusBottomRight() {
            return this.aq;
        }
        set iconCornerRadiusBottomRight(a) {
            this.aq = a;
            this.e("IconCornerRadiusBottomRight");
        }
        get iconCornerRadiusBottomLeft() {
            return this.ap;
        }
        set iconCornerRadiusBottomLeft(a) {
            this.ap = a;
            this.e("IconCornerRadiusBottomLeft");
        }
        get iconCornerRadiusTopLeft() {
            return this.ar;
        }
        set iconCornerRadiusTopLeft(a) {
            this.ar = a;
            this.e("IconCornerRadiusTopLeft");
        }
        get iconCornerRadiusTopRight() {
            return this.as;
        }
        set iconCornerRadiusTopRight(a) {
            this.as = a;
            this.e("IconCornerRadiusTopRight");
        }
        get iconRippleColor() {
            return this.fr;
        }
        set iconRippleColor(a) {
            this.fr = a;
            this.e("IconRippleColor");
        }
        get iconTextColor() {
            return this.fs;
        }
        set iconTextColor(a) {
            this.fs = a;
            this.e("IconTextColor");
        }
        get iconHoverTextColor() {
            return this.fq;
        }
        set iconHoverTextColor(a) {
            this.fq = a;
            this.e("IconHoverTextColor");
        }
        get iconFocusBackgroundColor() {
            return this.fn;
        }
        set iconFocusBackgroundColor(a) {
            this.fn = a;
            this.e("IconFocusBackgroundColor");
        }
        get iconFocusTextColor() {
            return this.fo;
        }
        set iconFocusTextColor(a) {
            this.fo = a;
            this.e("IconFocusTextColor");
        }
        get inputId() {
            return this.fu;
        }
        set inputId(a) {
            this.fu = a;
            this.e("InputId");
        }
        get id() {
            return this.ft;
        }
        set id(a) {
            this.ft = a;
            this.e("Id");
        }
        get display() {
            return this.ek;
        }
        set display(a) {
            this.ek = a;
            this.e("Display");
        }
        get flexDirection() {
            return this.e9;
        }
        set flexDirection(a) {
            this.e9 = a;
            this.e("FlexDirection");
        }
        get flexGrow() {
            return this.fa;
        }
        set flexGrow(a) {
            this.fa = a;
            this.e("FlexGrow");
        }
        get alignItems() {
            return this.ea;
        }
        set alignItems(a) {
            this.ea = a;
            this.e("AlignItems");
        }
        get valueRef() {
            return this.gk;
        }
        set valueRef(a) {
            this.gk = a;
            this.e("ValueRef");
        }
        get name() {
            return this.fv;
        }
        set name(a) {
            this.fv = a;
            this.e("Name");
        }
        get tabIndex() {
            return this.cz;
        }
        set tabIndex(a) {
            this.cz = a;
            this.e("TabIndex");
        }
        get focused() {
            return this.j;
        }
        set focused(a) {
            this.j = a;
            this.e("Focused");
        }
        get isFocusStyleEnabled() {
            return this.k;
        }
        set isFocusStyleEnabled(a) {
            this.k = a;
            this.e("IsFocusStyleEnabled");
        }
        get ariaLabel() {
            return this.eb;
        }
        set ariaLabel(a) {
            this.eb = a;
            this.e("AriaLabel");
        }
        get disableRipple() {
            return this.h;
        }
        set disableRipple(a) {
            this.h = a;
            this.e("DisableRipple");
        }
        get actualDisableRipple() {
            return this.f;
        }
        set actualDisableRipple(a) {
            this.f = a;
            this.e("ActualDisableRipple");
        }
        get isHover() {
            return this.l;
        }
        set isHover(a) {
            this.l = a;
            this.e("IsHover");
        }
        get disabled() {
            return this.g;
        }
        set disabled(a) {
            this.g = a;
            this.e("Disabled");
        }
        get disableTransitions() {
            return this.i;
        }
        set disableTransitions(a) {
            this.i = a;
            this.e("DisableTransitions");
        }
        get clickedRef() {
            return this.ef;
        }
        set clickedRef(a) {
            this.ef = a;
            this.e("ClickedRef");
        }
    }
    ButtonDescription.$t = /*@__PURE__*/ markType(ButtonDescription, 'ButtonDescription', Description.$);
    return ButtonDescription;
})();
