// igniteui-react-core/esm2015/lib/CellKeyDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CellKeyDescription = /*@__PURE__*/ (() => {
    class CellKeyDescription extends Description {
        constructor() {
            super();
            this.f = null;
            this.i = null;
            this.l = null;
        }
        get_type() {
            return "CellKey";
        }
        get type() {
            return this.get_type();
        }
        get primaryKey() {
            return this.f;
        }
        set primaryKey(a) {
            this.f = a;
            this.e("PrimaryKey");
        }
        get rowItem() {
            return this.i;
        }
        set rowItem(a) {
            this.i = a;
            this.e("RowItem");
        }
        get columnUniqueKey() {
            return this.l;
        }
        set columnUniqueKey(a) {
            this.l = a;
            this.e("ColumnUniqueKey");
        }
    }
    CellKeyDescription.$t = /*@__PURE__*/ markType(CellKeyDescription, 'CellKeyDescription', Description.$);
    CellKeyDescription.__marshalByValue = true;
    CellKeyDescription.__marshalByValueAlias = "CellKey";
    return CellKeyDescription;
})();
