// igniteui-react-core/esm2015/lib/CellRangeDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CellRangeDescription = /*@__PURE__*/ (() => {
    class CellRangeDescription extends Description {
        constructor() {
            super();
            this.i = 0;
            this.g = 0;
            this.j = 0;
            this.h = 0;
        }
        get_type() {
            return "CellRange";
        }
        get type() {
            return this.get_type();
        }
        get startColumn() {
            return this.i;
        }
        set startColumn(a) {
            this.i = a;
            this.e("StartColumn");
        }
        get endColumn() {
            return this.g;
        }
        set endColumn(a) {
            this.g = a;
            this.e("EndColumn");
        }
        get startRow() {
            return this.j;
        }
        set startRow(a) {
            this.j = a;
            this.e("StartRow");
        }
        get endRow() {
            return this.h;
        }
        set endRow(a) {
            this.h = a;
            this.e("EndRow");
        }
    }
    CellRangeDescription.$t = /*@__PURE__*/ markType(CellRangeDescription, 'CellRangeDescription', Description.$);
    CellRangeDescription.__marshalByValue = true;
    CellRangeDescription.__marshalByValueAlias = "CellRange";
    return CellRangeDescription;
})();
