// igniteui-react-core/esm2015/lib/DataSourceAggregatedResultDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataSourceAggregatedResultDescription = /*@__PURE__*/ (() => {
    class DataSourceAggregatedResultDescription extends Description {
        constructor() {
            super();
            this.j = null;
            this.k = null;
            this.f = null;
            this.l = null;
        }
        get_type() {
            return "DataSourceAggregatedResult";
        }
        get type() {
            return this.get_type();
        }
        get itemRef() {
            return this.j;
        }
        set itemRef(a) {
            this.j = a;
            this.e("ItemRef");
        }
        get transactionType() {
            return this.k;
        }
        set transactionType(a) {
            this.k = a;
            this.e("TransactionType");
        }
        get keys() {
            return this.f;
        }
        set keys(a) {
            this.f = a;
            this.e("Keys");
        }
        get valuesRef() {
            return this.l;
        }
        set valuesRef(a) {
            this.l = a;
            this.e("ValuesRef");
        }
    }
    DataSourceAggregatedResultDescription.$t = /*@__PURE__*/ markType(DataSourceAggregatedResultDescription, 'DataSourceAggregatedResultDescription', Description.$);
    DataSourceAggregatedResultDescription.__marshalByValue = true;
    DataSourceAggregatedResultDescription.__marshalByValueAlias = "DataSourceAggregatedResult";
    return DataSourceAggregatedResultDescription;
})();
