// igniteui-react-core/esm2015/lib/DateTimeFormatSpecifier.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FormatSpecifier } from "./FormatSpecifier";
import { markType } from "./type";
/**
 * @hidden
 */
export let DateTimeFormatSpecifier = /*@__PURE__*/ (() => {
    class DateTimeFormatSpecifier extends FormatSpecifier {
        constructor() {
            super();
            this.l = null;
            this.d = null;
            this.r = null;
            this.b = 0;
            this.c = null;
            this.f = null;
            this.p = null;
            this.m = null;
            this.s = null;
            this.j = null;
            this.k = null;
            this.h = null;
            this.u = null;
            this.g = null;
            this.v = null;
            this.o = null;
            this.e = null;
            this.i = null;
            this.n = null;
            this.q = null;
            this.t = null;
            this.b = -1;
        }
        a() {
            let a = {};
            if (this.d != null) {
                this.w(a, "dateStyle", this.d);
            }
            if (this.r != null) {
                this.w(a, "timeStyle", this.r);
            }
            if (this.b != -1) {
                this.w(a, "fractionalSecondDigits", this.b);
            }
            if (this.c != null) {
                this.w(a, "calendar", this.c);
            }
            if (this.f != null) {
                this.w(a, "dayPeriod", this.f);
            }
            if (this.p != null) {
                this.w(a, "numberingSystem", this.p);
            }
            if (this.m != null) {
                this.w(a, "localeMatcher", this.m);
            }
            if (this.s != null) {
                this.w(a, "timeZone", this.s);
            }
            if (this.j != null) {
                this.w(a, "hour12", this.j.toLowerCase() == "true" ? true : false);
            }
            if (this.k != null) {
                this.w(a, "hourCycle", this.k);
            }
            if (this.h != null) {
                this.w(a, "formatMatcher", this.h);
            }
            if (this.u != null) {
                this.w(a, "weekday", this.u);
            }
            if (this.g != null) {
                this.w(a, "era", this.g);
            }
            if (this.v != null) {
                this.w(a, "year", this.v);
            }
            if (this.o != null) {
                this.w(a, "month", this.o);
            }
            if (this.e != null) {
                this.w(a, "day", this.e);
            }
            if (this.i != null) {
                this.w(a, "hour", this.i);
            }
            if (this.n != null) {
                this.w(a, "minute", this.n);
            }
            if (this.q != null) {
                this.w(a, "second", this.q);
            }
            if (this.t != null) {
                this.w(a, "timeZoneName", this.t);
            }
            let locale_ = this.l;
            let options_ = a;
            return new Intl.DateTimeFormat(locale_, options_);
        }
        w(options_, propertyName_, value_) {
            options_[propertyName_] = value_;
        }
    }
    DateTimeFormatSpecifier.$t = /*@__PURE__*/ markType(DateTimeFormatSpecifier, 'DateTimeFormatSpecifier', FormatSpecifier.$);
    return DateTimeFormatSpecifier;
})();
