// igniteui-react-core/esm2015/lib/DefaultSummaryResult.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "./type";
import { ISummaryResult_$type } from "./ISummaryResult";
/**
 * @hidden
 */
export let DefaultSummaryResult = /*@__PURE__*/ (() => {
    class DefaultSummaryResult extends Base {
        constructor(a, b, c) {
            super();
            this.d = null;
            this.c = null;
            this.a = 0;
            this.b = 0;
            this.d = a;
            this.a = b;
            this.c = c;
        }
        get propertyName() {
            return this.d;
        }
        get operand() {
            return this.a;
        }
        get value() {
            return this.c;
        }
        get summaryIndex() {
            return this.b;
        }
        set summaryIndex(a) {
            this.b = a;
        }
    }
    DefaultSummaryResult.$t = /*@__PURE__*/ markType(DefaultSummaryResult, 'DefaultSummaryResult', Base.$, [ISummaryResult_$type]);
    return DefaultSummaryResult;
})();
