// igniteui-react-core/esm2015/lib/FunctionFilterExpressionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FilterExpressionDescription } from "./FilterExpressionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let FunctionFilterExpressionDescription = /*@__PURE__*/ (() => {
    class FunctionFilterExpressionDescription extends FilterExpressionDescription {
        constructor() {
            super();
            this.i = null;
            this.g = null;
        }
        get_type() {
            return "FunctionFilterExpression";
        }
        get functionType() {
            return this.i;
        }
        set functionType(a) {
            this.i = a;
            this.e("FunctionType");
        }
        get functionArguments() {
            return this.g;
        }
        set functionArguments(a) {
            this.g = a;
            this.e("FunctionArguments");
        }
    }
    FunctionFilterExpressionDescription.$t = /*@__PURE__*/ markType(FunctionFilterExpressionDescription, 'FunctionFilterExpressionDescription', FilterExpressionDescription.$);
    return FunctionFilterExpressionDescription;
})();
