// igniteui-react-core/esm2015/lib/GridCellEditEndedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridCellEditEndedEventArgsDescription = /*@__PURE__*/ (() => {
    class GridCellEditEndedEventArgsDescription extends Description {
        constructor() {
            super();
            this.f = null;
            this.h = 0;
            this.j = null;
        }
        get_type() {
            return "GridCellEditEndedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get column() {
            return this.f;
        }
        set column(a) {
            this.f = a;
            this.e("Column");
        }
        get row() {
            return this.h;
        }
        set row(a) {
            this.h = a;
            this.e("Row");
        }
        get itemRef() {
            return this.j;
        }
        set itemRef(a) {
            this.j = a;
            this.e("ItemRef");
        }
    }
    GridCellEditEndedEventArgsDescription.$t = /*@__PURE__*/ markType(GridCellEditEndedEventArgsDescription, 'GridCellEditEndedEventArgsDescription', Description.$);
    return GridCellEditEndedEventArgsDescription;
})();
