// igniteui-react-core/esm2015/lib/GridSelectedItemsChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridSelectedItemsChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class GridSelectedItemsChangedEventArgsDescription extends Description {
        constructor() {
            super();
            this.g = null;
            this.f = null;
            this.h = null;
        }
        get_type() {
            return "GridSelectedItemsChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get currentItems() {
            return this.g;
        }
        set currentItems(a) {
            this.g = a;
            this.e("CurrentItems");
        }
        get addedItems() {
            return this.f;
        }
        set addedItems(a) {
            this.f = a;
            this.e("AddedItems");
        }
        get removedItems() {
            return this.h;
        }
        set removedItems(a) {
            this.h = a;
            this.e("RemovedItems");
        }
    }
    GridSelectedItemsChangedEventArgsDescription.$t = /*@__PURE__*/ markType(GridSelectedItemsChangedEventArgsDescription, 'GridSelectedItemsChangedEventArgsDescription', Description.$);
    return GridSelectedItemsChangedEventArgsDescription;
})();
