// igniteui-react-core/esm2015/lib/GridCellEditEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridCellEditEventArgsDescription = /*@__PURE__*/ (() => {
    class GridCellEditEventArgsDescription extends Description {
        constructor() {
            super();
            this.f = null;
            this.l = null;
            this.k = null;
            this.j = null;
            this.h = false;
        }
        get_type() {
            return "GridCellEditEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get column() {
            return this.f;
        }
        set column(a) {
            this.f = a;
            this.e("Column");
        }
        get oldValueRef() {
            return this.l;
        }
        set oldValueRef(a) {
            this.l = a;
            this.e("OldValueRef");
        }
        get newValueRef() {
            return this.k;
        }
        set newValueRef(a) {
            this.k = a;
            this.e("NewValueRef");
        }
        get errorMessage() {
            return this.j;
        }
        set errorMessage(a) {
            this.j = a;
            this.e("ErrorMessage");
        }
        get cancel() {
            return this.h;
        }
        set cancel(a) {
            this.h = a;
            this.e("Cancel");
        }
    }
    GridCellEditEventArgsDescription.$t = /*@__PURE__*/ markType(GridCellEditEventArgsDescription, 'GridCellEditEventArgsDescription', Description.$);
    return GridCellEditEventArgsDescription;
})();
