// igniteui-react-core/esm2015/lib/GridSelectedCellsChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridSelectedCellsChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class GridSelectedCellsChangedEventArgsDescription extends Description {
        constructor() {
            super();
            this.g = null;
            this.f = null;
            this.h = null;
        }
        get_type() {
            return "GridSelectedCellsChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get currentCells() {
            return this.g;
        }
        set currentCells(a) {
            this.g = a;
            this.e("CurrentCells");
        }
        get addedCells() {
            return this.f;
        }
        set addedCells(a) {
            this.f = a;
            this.e("AddedCells");
        }
        get removedCells() {
            return this.h;
        }
        set removedCells(a) {
            this.h = a;
            this.e("RemovedCells");
        }
    }
    GridSelectedCellsChangedEventArgsDescription.$t = /*@__PURE__*/ markType(GridSelectedCellsChangedEventArgsDescription, 'GridSelectedCellsChangedEventArgsDescription', Description.$);
    return GridSelectedCellsChangedEventArgsDescription;
})();
