// igniteui-react-core/esm2015/lib/GridCommitCompletedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridCommitCompletedEventArgsDescription = /*@__PURE__*/ (() => {
    class GridCommitCompletedEventArgsDescription extends Description {
        constructor() {
            super();
            this.f = null;
            this.h = null;
        }
        get_type() {
            return "GridCommitCompletedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get transactionTypes() {
            return this.f;
        }
        set transactionTypes(a) {
            this.f = a;
            this.e("TransactionTypes");
        }
        get itemsRef() {
            return this.h;
        }
        set itemsRef(a) {
            this.h = a;
            this.e("ItemsRef");
        }
    }
    GridCommitCompletedEventArgsDescription.$t = /*@__PURE__*/ markType(GridCommitCompletedEventArgsDescription, 'GridCommitCompletedEventArgsDescription', Description.$);
    return GridCommitCompletedEventArgsDescription;
})();
