// igniteui-react-core/esm2015/lib/GridActiveCellChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridActiveCellChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class GridActiveCellChangedEventArgsDescription extends Description {
        constructor() {
            super();
            this.f = null;
            this.g = null;
        }
        get_type() {
            return "GridActiveCellChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get newActiveCell() {
            return this.f;
        }
        set newActiveCell(a) {
            this.f = a;
            this.e("NewActiveCell");
        }
        get oldActiveCell() {
            return this.g;
        }
        set oldActiveCell(a) {
            this.g = a;
            this.e("OldActiveCell");
        }
    }
    GridActiveCellChangedEventArgsDescription.$t = /*@__PURE__*/ markType(GridActiveCellChangedEventArgsDescription, 'GridActiveCellChangedEventArgsDescription', Description.$);
    return GridActiveCellChangedEventArgsDescription;
})();
