// igniteui-react-core/esm2015/lib/GridRowEditEndedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridRowEditEndedEventArgsDescription = /*@__PURE__*/ (() => {
    class GridRowEditEndedEventArgsDescription extends Description {
        constructor() {
            super();
            this.f = 0;
            this.h = null;
        }
        get_type() {
            return "GridRowEditEndedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get row() {
            return this.f;
        }
        set row(a) {
            this.f = a;
            this.e("Row");
        }
        get itemRef() {
            return this.h;
        }
        set itemRef(a) {
            this.h = a;
            this.e("ItemRef");
        }
    }
    GridRowEditEndedEventArgsDescription.$t = /*@__PURE__*/ markType(GridRowEditEndedEventArgsDescription, 'GridRowEditEndedEventArgsDescription', Description.$);
    return GridRowEditEndedEventArgsDescription;
})();
