// igniteui-react-core/esm2015/lib/GridColumnOptionsSimpleSectionBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GridColumnOptionsSectionBaseDescription } from "./GridColumnOptionsSectionBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GridColumnOptionsSimpleSectionBaseDescription = /*@__PURE__*/ (() => {
    class GridColumnOptionsSimpleSectionBaseDescription extends GridColumnOptionsSectionBaseDescription {
        constructor() {
            super();
            this.as = null;
            this.aq = false;
        }
        get_type() {
            return "GridColumnOptionsSimpleSectionBase";
        }
        get caption() {
            return this.as;
        }
        set caption(a) {
            this.as = a;
            this.e("Caption");
        }
        get showCaption() {
            return this.aq;
        }
        set showCaption(a) {
            this.aq = a;
            this.e("ShowCaption");
        }
    }
    GridColumnOptionsSimpleSectionBaseDescription.$t = /*@__PURE__*/ markType(GridColumnOptionsSimpleSectionBaseDescription, 'GridColumnOptionsSimpleSectionBaseDescription', GridColumnOptionsSectionBaseDescription.$);
    return GridColumnOptionsSimpleSectionBaseDescription;
})();
