// igniteui-react-core/esm2015/lib/igr-shape-filter-record-event-args.js
import { IgrShapefileRecord } from "./igr-shapefile-record";
import { ShapeFilterRecordEventArgs as ShapeFilterRecordEventArgs_internal } from "./ShapeFilterRecordEventArgs";
import { ensureBool } from "./componentUtil";
export class IgrShapeFilterRecordEventArgs {
    createImplementation() {
        return new ShapeFilterRecordEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    get record() {
        if (this.i.a == null) {
            return null;
        }
        if (!this.i.a.externalObject) {
            let e = new IgrShapefileRecord();
            e._implementation = this.i.a;
            this.i.a.externalObject = e;
        }
        return this.i.a.externalObject;
    }
    set record(v) {
        v == null ? this.i.a = null : this.i.a = v.i;
    }
    get shouldInclude() {
        return this.i.b;
    }
    set shouldInclude(v) {
        this.i.b = ensureBool(v);
    }
}
