// igniteui-react-core/esm2015/lib/igr-date-time-format-specifier.js
import { IgrFormatSpecifier } from "./igr-format-specifier";
import { DateTimeFormatSpecifier as DateTimeFormatSpecifier_internal } from "./DateTimeFormatSpecifier";
export class IgrDateTimeFormatSpecifier extends IgrFormatSpecifier {
    createImplementation() {
        return new DateTimeFormatSpecifier_internal();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    get locale() {
        return this.i.l;
    }
    set locale(v) {
        this.i.l = v;
    }
    get dateStyle() {
        return this.i.d;
    }
    set dateStyle(v) {
        this.i.d = v;
    }
    get timeStyle() {
        return this.i.r;
    }
    set timeStyle(v) {
        this.i.r = v;
    }
    get fractionalSecondDigits() {
        return this.i.b;
    }
    set fractionalSecondDigits(v) {
        this.i.b = +v;
    }
    get calendar() {
        return this.i.c;
    }
    set calendar(v) {
        this.i.c = v;
    }
    get dayPeriod() {
        return this.i.f;
    }
    set dayPeriod(v) {
        this.i.f = v;
    }
    get numberingSystem() {
        return this.i.p;
    }
    set numberingSystem(v) {
        this.i.p = v;
    }
    get localeMatcher() {
        return this.i.m;
    }
    set localeMatcher(v) {
        this.i.m = v;
    }
    get timeZone() {
        return this.i.s;
    }
    set timeZone(v) {
        this.i.s = v;
    }
    get hour12() {
        return this.i.j;
    }
    set hour12(v) {
        this.i.j = v;
    }
    get hourCycle() {
        return this.i.k;
    }
    set hourCycle(v) {
        this.i.k = v;
    }
    get formatMatcher() {
        return this.i.h;
    }
    set formatMatcher(v) {
        this.i.h = v;
    }
    get weekDay() {
        return this.i.u;
    }
    set weekDay(v) {
        this.i.u = v;
    }
    get era() {
        return this.i.g;
    }
    set era(v) {
        this.i.g = v;
    }
    get year() {
        return this.i.v;
    }
    set year(v) {
        this.i.v = v;
    }
    get month() {
        return this.i.o;
    }
    set month(v) {
        this.i.o = v;
    }
    get day() {
        return this.i.e;
    }
    set day(v) {
        this.i.e = v;
    }
    get hour() {
        return this.i.i;
    }
    set hour(v) {
        this.i.i = v;
    }
    get minute() {
        return this.i.n;
    }
    set minute(v) {
        this.i.n = v;
    }
    get second() {
        return this.i.q;
    }
    set second(v) {
        this.i.q = v;
    }
    get timeZoneName() {
        return this.i.t;
    }
    set timeZoneName(v) {
        this.i.t = v;
    }
}
