// igniteui-react-core/esm2015/lib/igr-number-format-specifier.js
import { IgrFormatSpecifier } from "./igr-format-specifier";
import { NumberFormatSpecifier as NumberFormatSpecifier_internal } from "./NumberFormatSpecifier";
import { ensureBool } from "./componentUtil";
export class IgrNumberFormatSpecifier extends IgrFormatSpecifier {
    createImplementation() {
        return new NumberFormatSpecifier_internal();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor() {
        super();
    }
    get locale() {
        return this.i.l;
    }
    set locale(v) {
        this.i.l = v;
    }
    get compactDisplay() {
        return this.i.h;
    }
    set compactDisplay(v) {
        this.i.h = v;
    }
    get currency() {
        return this.i.i;
    }
    set currency(v) {
        this.i.i = v;
    }
    get currencyDisplay() {
        return this.i.j;
    }
    set currencyDisplay(v) {
        this.i.j = v;
    }
    get currencySign() {
        return this.i.k;
    }
    set currencySign(v) {
        this.i.k = v;
    }
    get localeMatcher() {
        return this.i.m;
    }
    set localeMatcher(v) {
        this.i.m = v;
    }
    get notation() {
        return this.i.n;
    }
    set notation(v) {
        this.i.n = v;
    }
    get numberingSystem() {
        return this.i.o;
    }
    set numberingSystem(v) {
        this.i.o = v;
    }
    get signDisplay() {
        return this.i.p;
    }
    set signDisplay(v) {
        this.i.p = v;
    }
    get style() {
        return this.i.q;
    }
    set style(v) {
        this.i.q = v;
    }
    get unit() {
        return this.i.r;
    }
    set unit(v) {
        this.i.r = v;
    }
    get unitDisplay() {
        return this.i.s;
    }
    set unitDisplay(v) {
        this.i.s = v;
    }
    get useGrouping() {
        return this.i.b;
    }
    set useGrouping(v) {
        this.i.b = ensureBool(v);
    }
    get minimumIntegerDigits() {
        return this.i.f;
    }
    set minimumIntegerDigits(v) {
        this.i.f = +v;
    }
    get minimumFractionDigits() {
        return this.i.e;
    }
    set minimumFractionDigits(v) {
        this.i.e = +v;
    }
    get maximumFractionDigits() {
        return this.i.c;
    }
    set maximumFractionDigits(v) {
        this.i.c = +v;
    }
    get minimumSignificantDigits() {
        return this.i.g;
    }
    set minimumSignificantDigits(v) {
        this.i.g = +v;
    }
    get maximumSignificantDigits() {
        return this.i.d;
    }
    set maximumSignificantDigits(v) {
        this.i.d = +v;
    }
}
