// igniteui-react-core/esm2015/lib/InputGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let InputGroupDescription = /*@__PURE__*/ (() => {
    class InputGroupDescription extends Description {
        constructor() {
            super();
            this.gs = null;
            this.gr = null;
            this.f4 = null;
            this.gb = null;
            this.f3 = null;
            this.f5 = null;
            this.r = 0;
            this.aa = 0;
            this.x = 0;
            this.w = 0;
            this.y = 0;
            this.z = 0;
            this.s = 0;
            this.t = 0;
            this.u = 0;
            this.v = 0;
            this.ai = 0;
            this.ah = 0;
            this.ae = 0;
            this.f8 = null;
            this.f9 = null;
            this.f7 = null;
            this.f6 = null;
            this.af = 0;
            this.ad = 0;
            this.ab = 0;
            this.ac = 0;
            this.ag = 0;
            this.f2 = null;
            this.a1 = 0;
            this.bs = 0;
            this.gc = null;
            this.bp = 0;
            this.bo = 0;
            this.bq = 0;
            this.br = 0;
            this.bk = 0;
            this.bl = 0;
            this.bm = 0;
            this.bn = 0;
            this.c0 = 0;
            this.cz = 0;
            this.bw = 0;
            this.hb = null;
            this.hc = null;
            this.gv = null;
            this.gu = null;
            this.cx = 0;
            this.bv = 0;
            this.bt = 0;
            this.bu = 0;
            this.cy = 0;
            this.ga = null;
            this.bx = 0;
            this.b6 = 0;
            this.gz = null;
            this.gy = null;
            this.b3 = 0;
            this.b2 = 0;
            this.b4 = 0;
            this.b5 = 0;
            this.by = 0;
            this.bz = 0;
            this.b0 = 0;
            this.b1 = 0;
            this.ce = 0;
            this.cd = 0;
            this.ca = 0;
            this.g2 = null;
            this.g3 = null;
            this.g1 = null;
            this.g0 = null;
            this.cb = 0;
            this.b9 = 0;
            this.b7 = 0;
            this.b8 = 0;
            this.cc = 0;
            this.gx = null;
            this.a2 = 0;
            this.bb = 0;
            this.gm = null;
            this.gl = null;
            this.a8 = 0;
            this.a7 = 0;
            this.a9 = 0;
            this.ba = 0;
            this.a3 = 0;
            this.a4 = 0;
            this.a5 = 0;
            this.a6 = 0;
            this.bj = 0;
            this.bi = 0;
            this.bf = 0;
            this.gp = null;
            this.gq = null;
            this.go = null;
            this.gn = null;
            this.bg = 0;
            this.be = 0;
            this.bc = 0;
            this.bd = 0;
            this.bh = 0;
            this.gk = null;
            this.aj = 0;
            this.as = 0;
            this.gf = null;
            this.ge = null;
            this.ap = 0;
            this.ao = 0;
            this.aq = 0;
            this.ar = 0;
            this.ak = 0;
            this.al = 0;
            this.am = 0;
            this.an = 0;
            this.a0 = 0;
            this.az = 0;
            this.aw = 0;
            this.gi = null;
            this.gj = null;
            this.gh = null;
            this.gg = null;
            this.ax = 0;
            this.av = 0;
            this.at = 0;
            this.au = 0;
            this.ay = 0;
            this.gd = null;
            this.cf = 0;
            this.co = 0;
            this.g6 = null;
            this.g5 = null;
            this.cl = 0;
            this.ck = 0;
            this.cm = 0;
            this.cn = 0;
            this.cg = 0;
            this.ch = 0;
            this.ci = 0;
            this.cj = 0;
            this.cw = 0;
            this.cv = 0;
            this.cs = 0;
            this.g9 = null;
            this.ha = null;
            this.g8 = null;
            this.g7 = null;
            this.ct = 0;
            this.cr = 0;
            this.cp = 0;
            this.cq = 0;
            this.cu = 0;
            this.g4 = null;
            this.j = false;
            this.l = false;
            this.k = false;
            this.h = false;
            this.gt = null;
            this.gw = null;
            this.i = false;
            this.f = null;
        }
        get_type() {
            return "InputGroup";
        }
        get type() {
            return this.get_type();
        }
        get displayType() {
            return this.gs;
        }
        set displayType(a) {
            this.gs = a;
            this.e("DisplayType");
        }
        get density() {
            return this.gr;
        }
        set density(a) {
            this.gr = a;
            this.e("Density");
        }
        get actualDensity() {
            return this.f4;
        }
        set actualDensity(a) {
            this.f4 = a;
            this.e("ActualDensity");
        }
        get baseTheme() {
            return this.gb;
        }
        set baseTheme(a) {
            this.gb = a;
            this.e("BaseTheme");
        }
        get actualBorderColor() {
            return this.f3;
        }
        set actualBorderColor(a) {
            this.f3 = a;
            this.e("ActualBorderColor");
        }
        get actualFocusBorderColor() {
            return this.f5;
        }
        set actualFocusBorderColor(a) {
            this.f5 = a;
            this.e("ActualFocusBorderColor");
        }
        get actualBorderWidth() {
            return this.r;
        }
        set actualBorderWidth(a) {
            this.r = a;
            this.e("ActualBorderWidth");
        }
        get actualFocusBorderWidth() {
            return this.aa;
        }
        set actualFocusBorderWidth(a) {
            this.aa = a;
            this.e("ActualFocusBorderWidth");
        }
        get actualCornerRadiusBottomRight() {
            return this.x;
        }
        set actualCornerRadiusBottomRight(a) {
            this.x = a;
            this.e("ActualCornerRadiusBottomRight");
        }
        get actualCornerRadiusBottomLeft() {
            return this.w;
        }
        set actualCornerRadiusBottomLeft(a) {
            this.w = a;
            this.e("ActualCornerRadiusBottomLeft");
        }
        get actualCornerRadiusTopLeft() {
            return this.y;
        }
        set actualCornerRadiusTopLeft(a) {
            this.y = a;
            this.e("ActualCornerRadiusTopLeft");
        }
        get actualCornerRadiusTopRight() {
            return this.z;
        }
        set actualCornerRadiusTopRight(a) {
            this.z = a;
            this.e("ActualCornerRadiusTopRight");
        }
        get actualContentPaddingBottom() {
            return this.s;
        }
        set actualContentPaddingBottom(a) {
            this.s = a;
            this.e("ActualContentPaddingBottom");
        }
        get actualContentPaddingLeft() {
            return this.t;
        }
        set actualContentPaddingLeft(a) {
            this.t = a;
            this.e("ActualContentPaddingLeft");
        }
        get actualContentPaddingRight() {
            return this.u;
        }
        set actualContentPaddingRight(a) {
            this.u = a;
            this.e("ActualContentPaddingRight");
        }
        get actualContentPaddingTop() {
            return this.v;
        }
        set actualContentPaddingTop(a) {
            this.v = a;
            this.e("ActualContentPaddingTop");
        }
        get actualUnderlineWidth() {
            return this.ai;
        }
        set actualUnderlineWidth(a) {
            this.ai = a;
            this.e("ActualUnderlineWidth");
        }
        get actualUnderlineRippleWidth() {
            return this.ah;
        }
        set actualUnderlineRippleWidth(a) {
            this.ah = a;
            this.e("ActualUnderlineRippleWidth");
        }
        get actualHoverUnderlineWidth() {
            return this.ae;
        }
        set actualHoverUnderlineWidth(a) {
            this.ae = a;
            this.e("ActualHoverUnderlineWidth");
        }
        get actualUnderlineColor() {
            return this.f8;
        }
        set actualUnderlineColor(a) {
            this.f8 = a;
            this.e("ActualUnderlineColor");
        }
        get actualUnderlineRippleColor() {
            return this.f9;
        }
        set actualUnderlineRippleColor(a) {
            this.f9 = a;
            this.e("ActualUnderlineRippleColor");
        }
        get actualHoverUnderlineColor() {
            return this.f7;
        }
        set actualHoverUnderlineColor(a) {
            this.f7 = a;
            this.e("ActualHoverUnderlineColor");
        }
        get actualFocusUnderlineColor() {
            return this.f6;
        }
        set actualFocusUnderlineColor(a) {
            this.f6 = a;
            this.e("ActualFocusUnderlineColor");
        }
        get actualUnderlineOpacity() {
            return this.af;
        }
        set actualUnderlineOpacity(a) {
            this.af = a;
            this.e("ActualUnderlineOpacity");
        }
        get actualHoverUnderlineOpacity() {
            return this.ad;
        }
        set actualHoverUnderlineOpacity(a) {
            this.ad = a;
            this.e("ActualHoverUnderlineOpacity");
        }
        get actualFocusUnderlineOpacity() {
            return this.ab;
        }
        set actualFocusUnderlineOpacity(a) {
            this.ab = a;
            this.e("ActualFocusUnderlineOpacity");
        }
        get actualFocusUnderlineRippleOpacity() {
            return this.ac;
        }
        set actualFocusUnderlineRippleOpacity(a) {
            this.ac = a;
            this.e("ActualFocusUnderlineRippleOpacity");
        }
        get actualUnderlineRippleOpacity() {
            return this.ag;
        }
        set actualUnderlineRippleOpacity(a) {
            this.ag = a;
            this.e("ActualUnderlineRippleOpacity");
        }
        get actualBackgroundColor() {
            return this.f2;
        }
        set actualBackgroundColor(a) {
            this.f2 = a;
            this.e("ActualBackgroundColor");
        }
        get borderWidth() {
            return this.a1;
        }
        set borderWidth(a) {
            this.a1 = a;
            this.e("BorderWidth");
        }
        get focusBorderWidth() {
            return this.bs;
        }
        set focusBorderWidth(a) {
            this.bs = a;
            this.e("FocusBorderWidth");
        }
        get borderColor() {
            return this.gc;
        }
        set borderColor(a) {
            this.gc = a;
            this.e("BorderColor");
        }
        get cornerRadiusBottomRight() {
            return this.bp;
        }
        set cornerRadiusBottomRight(a) {
            this.bp = a;
            this.e("CornerRadiusBottomRight");
        }
        get cornerRadiusBottomLeft() {
            return this.bo;
        }
        set cornerRadiusBottomLeft(a) {
            this.bo = a;
            this.e("CornerRadiusBottomLeft");
        }
        get cornerRadiusTopLeft() {
            return this.bq;
        }
        set cornerRadiusTopLeft(a) {
            this.bq = a;
            this.e("CornerRadiusTopLeft");
        }
        get cornerRadiusTopRight() {
            return this.br;
        }
        set cornerRadiusTopRight(a) {
            this.br = a;
            this.e("CornerRadiusTopRight");
        }
        get contentPaddingBottom() {
            return this.bk;
        }
        set contentPaddingBottom(a) {
            this.bk = a;
            this.e("ContentPaddingBottom");
        }
        get contentPaddingLeft() {
            return this.bl;
        }
        set contentPaddingLeft(a) {
            this.bl = a;
            this.e("ContentPaddingLeft");
        }
        get contentPaddingRight() {
            return this.bm;
        }
        set contentPaddingRight(a) {
            this.bm = a;
            this.e("ContentPaddingRight");
        }
        get contentPaddingTop() {
            return this.bn;
        }
        set contentPaddingTop(a) {
            this.bn = a;
            this.e("ContentPaddingTop");
        }
        get underlineWidth() {
            return this.c0;
        }
        set underlineWidth(a) {
            this.c0 = a;
            this.e("UnderlineWidth");
        }
        get underlineRippleWidth() {
            return this.cz;
        }
        set underlineRippleWidth(a) {
            this.cz = a;
            this.e("UnderlineRippleWidth");
        }
        get hoverUnderlineWidth() {
            return this.bw;
        }
        set hoverUnderlineWidth(a) {
            this.bw = a;
            this.e("HoverUnderlineWidth");
        }
        get underlineColor() {
            return this.hb;
        }
        set underlineColor(a) {
            this.hb = a;
            this.e("UnderlineColor");
        }
        get underlineRippleColor() {
            return this.hc;
        }
        set underlineRippleColor(a) {
            this.hc = a;
            this.e("UnderlineRippleColor");
        }
        get hoverUnderlineColor() {
            return this.gv;
        }
        set hoverUnderlineColor(a) {
            this.gv = a;
            this.e("HoverUnderlineColor");
        }
        get focusUnderlineColor() {
            return this.gu;
        }
        set focusUnderlineColor(a) {
            this.gu = a;
            this.e("FocusUnderlineColor");
        }
        get underlineOpacity() {
            return this.cx;
        }
        set underlineOpacity(a) {
            this.cx = a;
            this.e("UnderlineOpacity");
        }
        get hoverUnderlineOpacity() {
            return this.bv;
        }
        set hoverUnderlineOpacity(a) {
            this.bv = a;
            this.e("HoverUnderlineOpacity");
        }
        get focusUnderlineOpacity() {
            return this.bt;
        }
        set focusUnderlineOpacity(a) {
            this.bt = a;
            this.e("FocusUnderlineOpacity");
        }
        get focusUnderlineRippleOpacity() {
            return this.bu;
        }
        set focusUnderlineRippleOpacity(a) {
            this.bu = a;
            this.e("FocusUnderlineRippleOpacity");
        }
        get underlineRippleOpacity() {
            return this.cy;
        }
        set underlineRippleOpacity(a) {
            this.cy = a;
            this.e("UnderlineRippleOpacity");
        }
        get backgroundColor() {
            return this.ga;
        }
        set backgroundColor(a) {
            this.ga = a;
            this.e("BackgroundColor");
        }
        get lineTypeBorderWidth() {
            return this.bx;
        }
        set lineTypeBorderWidth(a) {
            this.bx = a;
            this.e("LineTypeBorderWidth");
        }
        get lineTypeFocusBorderWidth() {
            return this.b6;
        }
        set lineTypeFocusBorderWidth(a) {
            this.b6 = a;
            this.e("LineTypeFocusBorderWidth");
        }
        get lineTypeFocusBorderColor() {
            return this.gz;
        }
        set lineTypeFocusBorderColor(a) {
            this.gz = a;
            this.e("LineTypeFocusBorderColor");
        }
        get lineTypeBorderColor() {
            return this.gy;
        }
        set lineTypeBorderColor(a) {
            this.gy = a;
            this.e("LineTypeBorderColor");
        }
        get lineTypeCornerRadiusBottomRight() {
            return this.b3;
        }
        set lineTypeCornerRadiusBottomRight(a) {
            this.b3 = a;
            this.e("LineTypeCornerRadiusBottomRight");
        }
        get lineTypeCornerRadiusBottomLeft() {
            return this.b2;
        }
        set lineTypeCornerRadiusBottomLeft(a) {
            this.b2 = a;
            this.e("LineTypeCornerRadiusBottomLeft");
        }
        get lineTypeCornerRadiusTopLeft() {
            return this.b4;
        }
        set lineTypeCornerRadiusTopLeft(a) {
            this.b4 = a;
            this.e("LineTypeCornerRadiusTopLeft");
        }
        get lineTypeCornerRadiusTopRight() {
            return this.b5;
        }
        set lineTypeCornerRadiusTopRight(a) {
            this.b5 = a;
            this.e("LineTypeCornerRadiusTopRight");
        }
        get lineTypeContentPaddingBottom() {
            return this.by;
        }
        set lineTypeContentPaddingBottom(a) {
            this.by = a;
            this.e("LineTypeContentPaddingBottom");
        }
        get lineTypeContentPaddingLeft() {
            return this.bz;
        }
        set lineTypeContentPaddingLeft(a) {
            this.bz = a;
            this.e("LineTypeContentPaddingLeft");
        }
        get lineTypeContentPaddingRight() {
            return this.b0;
        }
        set lineTypeContentPaddingRight(a) {
            this.b0 = a;
            this.e("LineTypeContentPaddingRight");
        }
        get lineTypeContentPaddingTop() {
            return this.b1;
        }
        set lineTypeContentPaddingTop(a) {
            this.b1 = a;
            this.e("LineTypeContentPaddingTop");
        }
        get lineTypeUnderlineWidth() {
            return this.ce;
        }
        set lineTypeUnderlineWidth(a) {
            this.ce = a;
            this.e("LineTypeUnderlineWidth");
        }
        get lineTypeUnderlineRippleWidth() {
            return this.cd;
        }
        set lineTypeUnderlineRippleWidth(a) {
            this.cd = a;
            this.e("LineTypeUnderlineRippleWidth");
        }
        get lineTypeHoverUnderlineWidth() {
            return this.ca;
        }
        set lineTypeHoverUnderlineWidth(a) {
            this.ca = a;
            this.e("LineTypeHoverUnderlineWidth");
        }
        get lineTypeUnderlineColor() {
            return this.g2;
        }
        set lineTypeUnderlineColor(a) {
            this.g2 = a;
            this.e("LineTypeUnderlineColor");
        }
        get lineTypeUnderlineRippleColor() {
            return this.g3;
        }
        set lineTypeUnderlineRippleColor(a) {
            this.g3 = a;
            this.e("LineTypeUnderlineRippleColor");
        }
        get lineTypeHoverUnderlineColor() {
            return this.g1;
        }
        set lineTypeHoverUnderlineColor(a) {
            this.g1 = a;
            this.e("LineTypeHoverUnderlineColor");
        }
        get lineTypeFocusUnderlineColor() {
            return this.g0;
        }
        set lineTypeFocusUnderlineColor(a) {
            this.g0 = a;
            this.e("LineTypeFocusUnderlineColor");
        }
        get lineTypeUnderlineOpacity() {
            return this.cb;
        }
        set lineTypeUnderlineOpacity(a) {
            this.cb = a;
            this.e("LineTypeUnderlineOpacity");
        }
        get lineTypeHoverUnderlineOpacity() {
            return this.b9;
        }
        set lineTypeHoverUnderlineOpacity(a) {
            this.b9 = a;
            this.e("LineTypeHoverUnderlineOpacity");
        }
        get lineTypeFocusUnderlineOpacity() {
            return this.b7;
        }
        set lineTypeFocusUnderlineOpacity(a) {
            this.b7 = a;
            this.e("LineTypeFocusUnderlineOpacity");
        }
        get lineTypeFocusUnderlineRippleOpacity() {
            return this.b8;
        }
        set lineTypeFocusUnderlineRippleOpacity(a) {
            this.b8 = a;
            this.e("LineTypeFocusUnderlineRippleOpacity");
        }
        get lineTypeUnderlineRippleOpacity() {
            return this.cc;
        }
        set lineTypeUnderlineRippleOpacity(a) {
            this.cc = a;
            this.e("LineTypeUnderlineRippleOpacity");
        }
        get lineTypeBackgroundColor() {
            return this.gx;
        }
        set lineTypeBackgroundColor(a) {
            this.gx = a;
            this.e("LineTypeBackgroundColor");
        }
        get boxTypeBorderWidth() {
            return this.a2;
        }
        set boxTypeBorderWidth(a) {
            this.a2 = a;
            this.e("BoxTypeBorderWidth");
        }
        get boxTypeFocusBorderWidth() {
            return this.bb;
        }
        set boxTypeFocusBorderWidth(a) {
            this.bb = a;
            this.e("BoxTypeFocusBorderWidth");
        }
        get boxTypeFocusBorderColor() {
            return this.gm;
        }
        set boxTypeFocusBorderColor(a) {
            this.gm = a;
            this.e("BoxTypeFocusBorderColor");
        }
        get boxTypeBorderColor() {
            return this.gl;
        }
        set boxTypeBorderColor(a) {
            this.gl = a;
            this.e("BoxTypeBorderColor");
        }
        get boxTypeCornerRadiusBottomRight() {
            return this.a8;
        }
        set boxTypeCornerRadiusBottomRight(a) {
            this.a8 = a;
            this.e("BoxTypeCornerRadiusBottomRight");
        }
        get boxTypeCornerRadiusBottomLeft() {
            return this.a7;
        }
        set boxTypeCornerRadiusBottomLeft(a) {
            this.a7 = a;
            this.e("BoxTypeCornerRadiusBottomLeft");
        }
        get boxTypeCornerRadiusTopLeft() {
            return this.a9;
        }
        set boxTypeCornerRadiusTopLeft(a) {
            this.a9 = a;
            this.e("BoxTypeCornerRadiusTopLeft");
        }
        get boxTypeCornerRadiusTopRight() {
            return this.ba;
        }
        set boxTypeCornerRadiusTopRight(a) {
            this.ba = a;
            this.e("BoxTypeCornerRadiusTopRight");
        }
        get boxTypeContentPaddingBottom() {
            return this.a3;
        }
        set boxTypeContentPaddingBottom(a) {
            this.a3 = a;
            this.e("BoxTypeContentPaddingBottom");
        }
        get boxTypeContentPaddingLeft() {
            return this.a4;
        }
        set boxTypeContentPaddingLeft(a) {
            this.a4 = a;
            this.e("BoxTypeContentPaddingLeft");
        }
        get boxTypeContentPaddingRight() {
            return this.a5;
        }
        set boxTypeContentPaddingRight(a) {
            this.a5 = a;
            this.e("BoxTypeContentPaddingRight");
        }
        get boxTypeContentPaddingTop() {
            return this.a6;
        }
        set boxTypeContentPaddingTop(a) {
            this.a6 = a;
            this.e("BoxTypeContentPaddingTop");
        }
        get boxTypeUnderlineWidth() {
            return this.bj;
        }
        set boxTypeUnderlineWidth(a) {
            this.bj = a;
            this.e("BoxTypeUnderlineWidth");
        }
        get boxTypeUnderlineRippleWidth() {
            return this.bi;
        }
        set boxTypeUnderlineRippleWidth(a) {
            this.bi = a;
            this.e("BoxTypeUnderlineRippleWidth");
        }
        get boxTypeHoverUnderlineWidth() {
            return this.bf;
        }
        set boxTypeHoverUnderlineWidth(a) {
            this.bf = a;
            this.e("BoxTypeHoverUnderlineWidth");
        }
        get boxTypeUnderlineColor() {
            return this.gp;
        }
        set boxTypeUnderlineColor(a) {
            this.gp = a;
            this.e("BoxTypeUnderlineColor");
        }
        get boxTypeUnderlineRippleColor() {
            return this.gq;
        }
        set boxTypeUnderlineRippleColor(a) {
            this.gq = a;
            this.e("BoxTypeUnderlineRippleColor");
        }
        get boxTypeHoverUnderlineColor() {
            return this.go;
        }
        set boxTypeHoverUnderlineColor(a) {
            this.go = a;
            this.e("BoxTypeHoverUnderlineColor");
        }
        get boxTypeFocusUnderlineColor() {
            return this.gn;
        }
        set boxTypeFocusUnderlineColor(a) {
            this.gn = a;
            this.e("BoxTypeFocusUnderlineColor");
        }
        get boxTypeUnderlineOpacity() {
            return this.bg;
        }
        set boxTypeUnderlineOpacity(a) {
            this.bg = a;
            this.e("BoxTypeUnderlineOpacity");
        }
        get boxTypeHoverUnderlineOpacity() {
            return this.be;
        }
        set boxTypeHoverUnderlineOpacity(a) {
            this.be = a;
            this.e("BoxTypeHoverUnderlineOpacity");
        }
        get boxTypeFocusUnderlineOpacity() {
            return this.bc;
        }
        set boxTypeFocusUnderlineOpacity(a) {
            this.bc = a;
            this.e("BoxTypeFocusUnderlineOpacity");
        }
        get boxTypeFocusUnderlineRippleOpacity() {
            return this.bd;
        }
        set boxTypeFocusUnderlineRippleOpacity(a) {
            this.bd = a;
            this.e("BoxTypeFocusUnderlineRippleOpacity");
        }
        get boxTypeUnderlineRippleOpacity() {
            return this.bh;
        }
        set boxTypeUnderlineRippleOpacity(a) {
            this.bh = a;
            this.e("BoxTypeUnderlineRippleOpacity");
        }
        get boxTypeBackgroundColor() {
            return this.gk;
        }
        set boxTypeBackgroundColor(a) {
            this.gk = a;
            this.e("BoxTypeBackgroundColor");
        }
        get borderTypeBorderWidth() {
            return this.aj;
        }
        set borderTypeBorderWidth(a) {
            this.aj = a;
            this.e("BorderTypeBorderWidth");
        }
        get borderTypeFocusBorderWidth() {
            return this.as;
        }
        set borderTypeFocusBorderWidth(a) {
            this.as = a;
            this.e("BorderTypeFocusBorderWidth");
        }
        get borderTypeFocusBorderColor() {
            return this.gf;
        }
        set borderTypeFocusBorderColor(a) {
            this.gf = a;
            this.e("BorderTypeFocusBorderColor");
        }
        get borderTypeBorderColor() {
            return this.ge;
        }
        set borderTypeBorderColor(a) {
            this.ge = a;
            this.e("BorderTypeBorderColor");
        }
        get borderTypeCornerRadiusBottomRight() {
            return this.ap;
        }
        set borderTypeCornerRadiusBottomRight(a) {
            this.ap = a;
            this.e("BorderTypeCornerRadiusBottomRight");
        }
        get borderTypeCornerRadiusBottomLeft() {
            return this.ao;
        }
        set borderTypeCornerRadiusBottomLeft(a) {
            this.ao = a;
            this.e("BorderTypeCornerRadiusBottomLeft");
        }
        get borderTypeCornerRadiusTopLeft() {
            return this.aq;
        }
        set borderTypeCornerRadiusTopLeft(a) {
            this.aq = a;
            this.e("BorderTypeCornerRadiusTopLeft");
        }
        get borderTypeCornerRadiusTopRight() {
            return this.ar;
        }
        set borderTypeCornerRadiusTopRight(a) {
            this.ar = a;
            this.e("BorderTypeCornerRadiusTopRight");
        }
        get borderTypeContentPaddingBottom() {
            return this.ak;
        }
        set borderTypeContentPaddingBottom(a) {
            this.ak = a;
            this.e("BorderTypeContentPaddingBottom");
        }
        get borderTypeContentPaddingLeft() {
            return this.al;
        }
        set borderTypeContentPaddingLeft(a) {
            this.al = a;
            this.e("BorderTypeContentPaddingLeft");
        }
        get borderTypeContentPaddingRight() {
            return this.am;
        }
        set borderTypeContentPaddingRight(a) {
            this.am = a;
            this.e("BorderTypeContentPaddingRight");
        }
        get borderTypeContentPaddingTop() {
            return this.an;
        }
        set borderTypeContentPaddingTop(a) {
            this.an = a;
            this.e("BorderTypeContentPaddingTop");
        }
        get borderTypeUnderlineWidth() {
            return this.a0;
        }
        set borderTypeUnderlineWidth(a) {
            this.a0 = a;
            this.e("BorderTypeUnderlineWidth");
        }
        get borderTypeUnderlineRippleWidth() {
            return this.az;
        }
        set borderTypeUnderlineRippleWidth(a) {
            this.az = a;
            this.e("BorderTypeUnderlineRippleWidth");
        }
        get borderTypeHoverUnderlineWidth() {
            return this.aw;
        }
        set borderTypeHoverUnderlineWidth(a) {
            this.aw = a;
            this.e("BorderTypeHoverUnderlineWidth");
        }
        get borderTypeUnderlineColor() {
            return this.gi;
        }
        set borderTypeUnderlineColor(a) {
            this.gi = a;
            this.e("BorderTypeUnderlineColor");
        }
        get borderTypeUnderlineRippleColor() {
            return this.gj;
        }
        set borderTypeUnderlineRippleColor(a) {
            this.gj = a;
            this.e("BorderTypeUnderlineRippleColor");
        }
        get borderTypeHoverUnderlineColor() {
            return this.gh;
        }
        set borderTypeHoverUnderlineColor(a) {
            this.gh = a;
            this.e("BorderTypeHoverUnderlineColor");
        }
        get borderTypeFocusUnderlineColor() {
            return this.gg;
        }
        set borderTypeFocusUnderlineColor(a) {
            this.gg = a;
            this.e("BorderTypeFocusUnderlineColor");
        }
        get borderTypeUnderlineOpacity() {
            return this.ax;
        }
        set borderTypeUnderlineOpacity(a) {
            this.ax = a;
            this.e("BorderTypeUnderlineOpacity");
        }
        get borderTypeHoverUnderlineOpacity() {
            return this.av;
        }
        set borderTypeHoverUnderlineOpacity(a) {
            this.av = a;
            this.e("BorderTypeHoverUnderlineOpacity");
        }
        get borderTypeFocusUnderlineOpacity() {
            return this.at;
        }
        set borderTypeFocusUnderlineOpacity(a) {
            this.at = a;
            this.e("BorderTypeFocusUnderlineOpacity");
        }
        get borderTypeFocusUnderlineRippleOpacity() {
            return this.au;
        }
        set borderTypeFocusUnderlineRippleOpacity(a) {
            this.au = a;
            this.e("BorderTypeFocusUnderlineRippleOpacity");
        }
        get borderTypeUnderlineRippleOpacity() {
            return this.ay;
        }
        set borderTypeUnderlineRippleOpacity(a) {
            this.ay = a;
            this.e("BorderTypeUnderlineRippleOpacity");
        }
        get borderTypeBackgroundColor() {
            return this.gd;
        }
        set borderTypeBackgroundColor(a) {
            this.gd = a;
            this.e("BorderTypeBackgroundColor");
        }
        get searchTypeBorderWidth() {
            return this.cf;
        }
        set searchTypeBorderWidth(a) {
            this.cf = a;
            this.e("SearchTypeBorderWidth");
        }
        get searchTypeFocusBorderWidth() {
            return this.co;
        }
        set searchTypeFocusBorderWidth(a) {
            this.co = a;
            this.e("SearchTypeFocusBorderWidth");
        }
        get searchTypeFocusBorderColor() {
            return this.g6;
        }
        set searchTypeFocusBorderColor(a) {
            this.g6 = a;
            this.e("SearchTypeFocusBorderColor");
        }
        get searchTypeBorderColor() {
            return this.g5;
        }
        set searchTypeBorderColor(a) {
            this.g5 = a;
            this.e("SearchTypeBorderColor");
        }
        get searchTypeCornerRadiusBottomRight() {
            return this.cl;
        }
        set searchTypeCornerRadiusBottomRight(a) {
            this.cl = a;
            this.e("SearchTypeCornerRadiusBottomRight");
        }
        get searchTypeCornerRadiusBottomLeft() {
            return this.ck;
        }
        set searchTypeCornerRadiusBottomLeft(a) {
            this.ck = a;
            this.e("SearchTypeCornerRadiusBottomLeft");
        }
        get searchTypeCornerRadiusTopLeft() {
            return this.cm;
        }
        set searchTypeCornerRadiusTopLeft(a) {
            this.cm = a;
            this.e("SearchTypeCornerRadiusTopLeft");
        }
        get searchTypeCornerRadiusTopRight() {
            return this.cn;
        }
        set searchTypeCornerRadiusTopRight(a) {
            this.cn = a;
            this.e("SearchTypeCornerRadiusTopRight");
        }
        get searchTypeContentPaddingBottom() {
            return this.cg;
        }
        set searchTypeContentPaddingBottom(a) {
            this.cg = a;
            this.e("SearchTypeContentPaddingBottom");
        }
        get searchTypeContentPaddingLeft() {
            return this.ch;
        }
        set searchTypeContentPaddingLeft(a) {
            this.ch = a;
            this.e("SearchTypeContentPaddingLeft");
        }
        get searchTypeContentPaddingRight() {
            return this.ci;
        }
        set searchTypeContentPaddingRight(a) {
            this.ci = a;
            this.e("SearchTypeContentPaddingRight");
        }
        get searchTypeContentPaddingTop() {
            return this.cj;
        }
        set searchTypeContentPaddingTop(a) {
            this.cj = a;
            this.e("SearchTypeContentPaddingTop");
        }
        get searchTypeUnderlineWidth() {
            return this.cw;
        }
        set searchTypeUnderlineWidth(a) {
            this.cw = a;
            this.e("SearchTypeUnderlineWidth");
        }
        get searchTypeUnderlineRippleWidth() {
            return this.cv;
        }
        set searchTypeUnderlineRippleWidth(a) {
            this.cv = a;
            this.e("SearchTypeUnderlineRippleWidth");
        }
        get searchTypeHoverUnderlineWidth() {
            return this.cs;
        }
        set searchTypeHoverUnderlineWidth(a) {
            this.cs = a;
            this.e("SearchTypeHoverUnderlineWidth");
        }
        get searchTypeUnderlineColor() {
            return this.g9;
        }
        set searchTypeUnderlineColor(a) {
            this.g9 = a;
            this.e("SearchTypeUnderlineColor");
        }
        get searchTypeUnderlineRippleColor() {
            return this.ha;
        }
        set searchTypeUnderlineRippleColor(a) {
            this.ha = a;
            this.e("SearchTypeUnderlineRippleColor");
        }
        get searchTypeHoverUnderlineColor() {
            return this.g8;
        }
        set searchTypeHoverUnderlineColor(a) {
            this.g8 = a;
            this.e("SearchTypeHoverUnderlineColor");
        }
        get searchTypeFocusUnderlineColor() {
            return this.g7;
        }
        set searchTypeFocusUnderlineColor(a) {
            this.g7 = a;
            this.e("SearchTypeFocusUnderlineColor");
        }
        get searchTypeUnderlineOpacity() {
            return this.ct;
        }
        set searchTypeUnderlineOpacity(a) {
            this.ct = a;
            this.e("SearchTypeUnderlineOpacity");
        }
        get searchTypeHoverUnderlineOpacity() {
            return this.cr;
        }
        set searchTypeHoverUnderlineOpacity(a) {
            this.cr = a;
            this.e("SearchTypeHoverUnderlineOpacity");
        }
        get searchTypeFocusUnderlineOpacity() {
            return this.cp;
        }
        set searchTypeFocusUnderlineOpacity(a) {
            this.cp = a;
            this.e("SearchTypeFocusUnderlineOpacity");
        }
        get searchTypeFocusUnderlineRippleOpacity() {
            return this.cq;
        }
        set searchTypeFocusUnderlineRippleOpacity(a) {
            this.cq = a;
            this.e("SearchTypeFocusUnderlineRippleOpacity");
        }
        get searchTypeUnderlineRippleOpacity() {
            return this.cu;
        }
        set searchTypeUnderlineRippleOpacity(a) {
            this.cu = a;
            this.e("SearchTypeUnderlineRippleOpacity");
        }
        get searchTypeBackgroundColor() {
            return this.g4;
        }
        set searchTypeBackgroundColor(a) {
            this.g4 = a;
            this.e("SearchTypeBackgroundColor");
        }
        get isExpanded() {
            return this.j;
        }
        set isExpanded(a) {
            this.j = a;
            this.e("IsExpanded");
        }
        get isHovered() {
            return this.l;
        }
        set isHovered(a) {
            this.l = a;
            this.e("IsHovered");
        }
        get isFocused() {
            return this.k;
        }
        set isFocused(a) {
            this.k = a;
            this.e("IsFocused");
        }
        get actualIsExpanded() {
            return this.h;
        }
        set actualIsExpanded(a) {
            this.h = a;
            this.e("ActualIsExpanded");
        }
        get focusBorderColor() {
            return this.gt;
        }
        set focusBorderColor(a) {
            this.gt = a;
            this.e("FocusBorderColor");
        }
        get id() {
            return this.gw;
        }
        set id(a) {
            this.gw = a;
            this.e("Id");
        }
        get disabled() {
            return this.i;
        }
        set disabled(a) {
            this.i = a;
            this.e("Disabled");
        }
        get inputs() {
            return this.f;
        }
        set inputs(a) {
            this.f = a;
            this.e("Inputs");
        }
    }
    InputGroupDescription.$t = /*@__PURE__*/ markType(InputGroupDescription, 'InputGroupDescription', Description.$);
    return InputGroupDescription;
})();
