// igniteui-react-core/esm2015/lib/LegendBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let LegendBaseDescription = /*@__PURE__*/ (() => {
    class LegendBaseDescription extends Description {
        constructor() {
            super();
            this.f = null;
            this.i = null;
            this.j = null;
            this.g = null;
            this.h = null;
            this.k = null;
        }
        get background() {
            return this.f;
        }
        set background(a) {
            this.f = a;
            this.e("Background");
        }
        get_type() {
            return "LegendBase";
        }
        get type() {
            return this.get_type();
        }
        get legendItemMouseLeftButtonDownRef() {
            return this.i;
        }
        set legendItemMouseLeftButtonDownRef(a) {
            this.i = a;
            this.e("LegendItemMouseLeftButtonDownRef");
        }
        get legendItemMouseLeftButtonUpRef() {
            return this.j;
        }
        set legendItemMouseLeftButtonUpRef(a) {
            this.j = a;
            this.e("LegendItemMouseLeftButtonUpRef");
        }
        get legendItemMouseEnterRef() {
            return this.g;
        }
        set legendItemMouseEnterRef(a) {
            this.g = a;
            this.e("LegendItemMouseEnterRef");
        }
        get legendItemMouseLeaveRef() {
            return this.h;
        }
        set legendItemMouseLeaveRef(a) {
            this.h = a;
            this.e("LegendItemMouseLeaveRef");
        }
        get legendItemMouseMoveRef() {
            return this.k;
        }
        set legendItemMouseMoveRef(a) {
            this.k = a;
            this.e("LegendItemMouseMoveRef");
        }
    }
    LegendBaseDescription.$t = /*@__PURE__*/ markType(LegendBaseDescription, 'LegendBaseDescription', Description.$);
    return LegendBaseDescription;
})();
