// igniteui-react-core/esm2015/lib/LegendMouseButtonEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let LegendMouseButtonEventArgsDescription = /*@__PURE__*/ (() => {
    class LegendMouseButtonEventArgsDescription extends Description {
        constructor() {
            super();
            this.f = false;
            this.h = null;
            this.i = null;
        }
        get_type() {
            return "LegendMouseButtonEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get handled() {
            return this.f;
        }
        set handled(a) {
            this.f = a;
            this.e("Handled");
        }
        get itemRef() {
            return this.h;
        }
        set itemRef(a) {
            this.h = a;
            this.e("ItemRef");
        }
        get legendItemRef() {
            return this.i;
        }
        set legendItemRef(a) {
            this.i = a;
            this.e("LegendItemRef");
        }
    }
    LegendMouseButtonEventArgsDescription.$t = /*@__PURE__*/ markType(LegendMouseButtonEventArgsDescription, 'LegendMouseButtonEventArgsDescription', Description.$);
    return LegendMouseButtonEventArgsDescription;
})();
