// igniteui-react-core/esm2015/lib/PieChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PieChartBaseDescription } from "./PieChartBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let PieChartDescription = /*@__PURE__*/ (() => {
    class PieChartDescription extends PieChartBaseDescription {
        constructor() {
            super();
            this.da = null;
            this.c9 = null;
            this.c8 = null;
        }
        get width() {
            return this.da;
        }
        set width(a) {
            this.da = a;
            this.e("Width");
        }
        get height() {
            return this.c9;
        }
        set height(a) {
            this.c9 = a;
            this.e("Height");
        }
        get background() {
            return this.c8;
        }
        set background(a) {
            this.c8 = a;
            this.e("Background");
        }
        get_type() {
            return "PieChart";
        }
    }
    PieChartDescription.$t = /*@__PURE__*/ markType(PieChartDescription, 'PieChartDescription', PieChartBaseDescription.$);
    return PieChartDescription;
})();
