// igniteui-react-core/esm2015/lib/PieSliceDataContextDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DataContextDescription } from "./DataContextDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let PieSliceDataContextDescription = /*@__PURE__*/ (() => {
    class PieSliceDataContextDescription extends DataContextDescription {
        constructor() {
            super();
            this.z = 0;
            this.x = false;
        }
        get_type() {
            return "PieSliceDataContext";
        }
        get percentValue() {
            return this.z;
        }
        set percentValue(a) {
            this.z = a;
            this.e("PercentValue");
        }
        get isOthersSlice() {
            return this.x;
        }
        set isOthersSlice(a) {
            this.x = a;
            this.e("IsOthersSlice");
        }
    }
    PieSliceDataContextDescription.$t = /*@__PURE__*/ markType(PieSliceDataContextDescription, 'PieSliceDataContextDescription', DataContextDescription.$);
    PieSliceDataContextDescription.__marshalByValue1 = true;
    PieSliceDataContextDescription.__marshalByValueAlias1 = "PieSliceDataContext";
    return PieSliceDataContextDescription;
})();
