// igniteui-react-core/esm2015/lib/RingSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let RingSeriesBaseDescription = /*@__PURE__*/ (() => {
    class RingSeriesBaseDescription extends Description {
        constructor() {
            super();
            this.bi = null;
            this.as = null;
            this.t = 0;
            this.au = null;
            this.at = null;
            this.ar = null;
            this.bj = null;
            this.az = null;
            this.a7 = null;
            this.a1 = null;
            this.a5 = null;
            this.a2 = null;
            this.a3 = null;
            this.x = 0;
            this.w = 0;
            this.a4 = null;
            this.v = 0;
            this.aa = 0;
            this.bd = null;
            this.bc = null;
            this.a9 = null;
            this.av = null;
            this.aw = null;
            this.ax = null;
            this.g = null;
            this.be = null;
            this.j = null;
            this.a6 = null;
            this.h = null;
            this.a8 = null;
            this.i = null;
            this.u = 0;
            this.ae = 0;
            this.ba = null;
            this.bb = null;
            this.z = 0;
            this.y = 0;
            this.bg = null;
            this.bh = null;
            this.ad = 0;
            this.ac = 0;
            this.f = null;
            this.k = null;
            this.a0 = null;
            this.ay = null;
            this.r = false;
            this.ab = 0;
            this.bf = null;
        }
        get tooltipTemplateRef() {
            return this.bi;
        }
        set tooltipTemplateRef(a) {
            this.bi = a;
            this.e("TooltipTemplateRef");
        }
        get fontFamily() {
            return this.as;
        }
        set fontFamily(a) {
            this.as = a;
            this.e("FontFamily");
        }
        get fontSize() {
            return this.t;
        }
        set fontSize(a) {
            this.t = a;
            this.e("FontSize");
        }
        get fontWeight() {
            return this.au;
        }
        set fontWeight(a) {
            this.au = a;
            this.e("FontWeight");
        }
        get fontStyle() {
            return this.at;
        }
        set fontStyle(a) {
            this.at = a;
            this.e("FontStyle");
        }
        get_type() {
            return "RingSeriesBase";
        }
        get type() {
            return this.get_type();
        }
        get dataSourceRef() {
            return this.ar;
        }
        set dataSourceRef(a) {
            this.ar = a;
            this.e("DataSourceRef");
        }
        get valueMemberPath() {
            return this.bj;
        }
        set valueMemberPath(a) {
            this.bj = a;
            this.e("ValueMemberPath");
        }
        get labelMemberPath() {
            return this.az;
        }
        set labelMemberPath(a) {
            this.az = a;
            this.e("LabelMemberPath");
        }
        get legendLabelMemberPath() {
            return this.a7;
        }
        set legendLabelMemberPath(a) {
            this.a7 = a;
            this.e("LegendLabelMemberPath");
        }
        get labelsPosition() {
            return this.a1;
        }
        set labelsPosition(a) {
            this.a1 = a;
            this.e("LabelsPosition");
        }
        get leaderLineVisibility() {
            return this.a5;
        }
        set leaderLineVisibility(a) {
            this.a5 = a;
            this.e("LeaderLineVisibility");
        }
        get leaderLineFill() {
            return this.a2;
        }
        set leaderLineFill(a) {
            this.a2 = a;
            this.e("LeaderLineFill");
        }
        get leaderLineStroke() {
            return this.a3;
        }
        set leaderLineStroke(a) {
            this.a3 = a;
            this.e("LeaderLineStroke");
        }
        get leaderLineStrokeThickness() {
            return this.x;
        }
        set leaderLineStrokeThickness(a) {
            this.x = a;
            this.e("LeaderLineStrokeThickness");
        }
        get leaderLineOpacity() {
            return this.w;
        }
        set leaderLineOpacity(a) {
            this.w = a;
            this.e("LeaderLineOpacity");
        }
        get leaderLineType() {
            return this.a4;
        }
        set leaderLineType(a) {
            this.a4 = a;
            this.e("LeaderLineType");
        }
        get leaderLineMargin() {
            return this.v;
        }
        set leaderLineMargin(a) {
            this.v = a;
            this.e("LeaderLineMargin");
        }
        get othersCategoryThreshold() {
            return this.aa;
        }
        set othersCategoryThreshold(a) {
            this.aa = a;
            this.e("OthersCategoryThreshold");
        }
        get othersCategoryType() {
            return this.bd;
        }
        set othersCategoryType(a) {
            this.bd = a;
            this.e("OthersCategoryType");
        }
        get othersCategoryText() {
            return this.bc;
        }
        set othersCategoryText(a) {
            this.bc = a;
            this.e("OthersCategoryText");
        }
        get legendRef() {
            return this.a9;
        }
        set legendRef(a) {
            this.a9 = a;
            this.e("LegendRef");
        }
        get formatLabelRef() {
            return this.av;
        }
        set formatLabelRef(a) {
            this.av = a;
            this.e("FormatLabelRef");
        }
        get formatLegendLabelRef() {
            return this.aw;
        }
        set formatLegendLabelRef(a) {
            this.aw = a;
            this.e("FormatLegendLabelRef");
        }
        get labelFormat() {
            return this.ax;
        }
        set labelFormat(a) {
            this.ax = a;
            this.e("LabelFormat");
        }
        get labelFormatSpecifiers() {
            return this.g;
        }
        set labelFormatSpecifiers(a) {
            this.g = a;
            this.e("LabelFormatSpecifiers");
        }
        get othersLabelFormat() {
            return this.be;
        }
        set othersLabelFormat(a) {
            this.be = a;
            this.e("OthersLabelFormat");
        }
        get othersLabelFormatSpecifiers() {
            return this.j;
        }
        set othersLabelFormatSpecifiers(a) {
            this.j = a;
            this.e("OthersLabelFormatSpecifiers");
        }
        get legendLabelFormat() {
            return this.a6;
        }
        set legendLabelFormat(a) {
            this.a6 = a;
            this.e("LegendLabelFormat");
        }
        get legendLabelFormatSpecifiers() {
            return this.h;
        }
        set legendLabelFormatSpecifiers(a) {
            this.h = a;
            this.e("LegendLabelFormatSpecifiers");
        }
        get legendOthersLabelFormat() {
            return this.a8;
        }
        set legendOthersLabelFormat(a) {
            this.a8 = a;
            this.e("LegendOthersLabelFormat");
        }
        get legendOthersLabelFormatSpecifiers() {
            return this.i;
        }
        set legendOthersLabelFormatSpecifiers(a) {
            this.i = a;
            this.e("LegendOthersLabelFormatSpecifiers");
        }
        get labelExtent() {
            return this.u;
        }
        set labelExtent(a) {
            this.u = a;
            this.e("LabelExtent");
        }
        get startAngle() {
            return this.ae;
        }
        set startAngle(a) {
            this.ae = a;
            this.e("StartAngle");
        }
        get othersCategoryFill() {
            return this.ba;
        }
        set othersCategoryFill(a) {
            this.ba = a;
            this.e("OthersCategoryFill");
        }
        get othersCategoryStroke() {
            return this.bb;
        }
        set othersCategoryStroke(a) {
            this.bb = a;
            this.e("OthersCategoryStroke");
        }
        get othersCategoryStrokeThickness() {
            return this.z;
        }
        set othersCategoryStrokeThickness(a) {
            this.z = a;
            this.e("OthersCategoryStrokeThickness");
        }
        get othersCategoryOpacity() {
            return this.y;
        }
        set othersCategoryOpacity(a) {
            this.y = a;
            this.e("OthersCategoryOpacity");
        }
        get selectedSliceFill() {
            return this.bg;
        }
        set selectedSliceFill(a) {
            this.bg = a;
            this.e("SelectedSliceFill");
        }
        get selectedSliceStroke() {
            return this.bh;
        }
        set selectedSliceStroke(a) {
            this.bh = a;
            this.e("SelectedSliceStroke");
        }
        get selectedSliceStrokeThickness() {
            return this.ad;
        }
        set selectedSliceStrokeThickness(a) {
            this.ad = a;
            this.e("SelectedSliceStrokeThickness");
        }
        get selectedSliceOpacity() {
            return this.ac;
        }
        set selectedSliceOpacity(a) {
            this.ac = a;
            this.e("SelectedSliceOpacity");
        }
        get brushes() {
            return this.f;
        }
        set brushes(a) {
            this.f = a;
            this.e("Brushes");
        }
        get outlines() {
            return this.k;
        }
        set outlines(a) {
            this.k = a;
            this.e("Outlines");
        }
        get labelOuterColor() {
            return this.a0;
        }
        set labelOuterColor(a) {
            this.a0 = a;
            this.e("LabelOuterColor");
        }
        get labelInnerColor() {
            return this.ay;
        }
        set labelInnerColor(a) {
            this.ay = a;
            this.e("LabelInnerColor");
        }
        get isSurfaceInteractionDisabled() {
            return this.r;
        }
        set isSurfaceInteractionDisabled(a) {
            this.r = a;
            this.e("IsSurfaceInteractionDisabled");
        }
        get radiusFactor() {
            return this.ab;
        }
        set radiusFactor(a) {
            this.ab = a;
            this.e("RadiusFactor");
        }
        get propertyUpdatedRef() {
            return this.bf;
        }
        set propertyUpdatedRef(a) {
            this.bf = a;
            this.e("PropertyUpdatedRef");
        }
    }
    RingSeriesBaseDescription.$t = /*@__PURE__*/ markType(RingSeriesBaseDescription, 'RingSeriesBaseDescription', Description.$);
    return RingSeriesBaseDescription;
})();
