// igniteui-react-core/esm2015/lib/SectionHeaderCellInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CellInfoDescription } from "./CellInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let SectionHeaderCellInfoDescription = /*@__PURE__*/ (() => {
    class SectionHeaderCellInfoDescription extends CellInfoDescription {
        constructor() {
            super();
            this.fc = null;
            this.fb = null;
            this.fd = null;
        }
        get_type() {
            return "SectionHeaderCellInfo";
        }
        get resolvedText() {
            return this.fc;
        }
        set resolvedText(a) {
            this.fc = a;
            this.e("ResolvedText");
        }
        get resolvedSummaryText() {
            return this.fb;
        }
        set resolvedSummaryText(a) {
            this.fb = a;
            this.e("ResolvedSummaryText");
        }
        get summaryDisplayMode() {
            return this.fd;
        }
        set summaryDisplayMode(a) {
            this.fd = a;
            this.e("SummaryDisplayMode");
        }
    }
    SectionHeaderCellInfoDescription.$t = /*@__PURE__*/ markType(SectionHeaderCellInfoDescription, 'SectionHeaderCellInfoDescription', CellInfoDescription.$);
    SectionHeaderCellInfoDescription.__marshalByValue1 = true;
    SectionHeaderCellInfoDescription.__marshalByValueAlias1 = "SectionHeaderCellModel";
    return SectionHeaderCellInfoDescription;
})();
