// igniteui-react-core/esm2015/lib/SelectedItemChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SelectedItemChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class SelectedItemChangedEventArgsDescription extends Description {
        constructor() {
            super();
            this.g = null;
            this.f = null;
        }
        get_type() {
            return "SelectedItemChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get oldItem() {
            return this.g;
        }
        set oldItem(a) {
            this.g = a;
            this.e("OldItem");
        }
        get newItem() {
            return this.f;
        }
        set newItem(a) {
            this.f = a;
            this.e("NewItem");
        }
    }
    SelectedItemChangedEventArgsDescription.$t = /*@__PURE__*/ markType(SelectedItemChangedEventArgsDescription, 'SelectedItemChangedEventArgsDescription', Description.$);
    return SelectedItemChangedEventArgsDescription;
})();
