// igniteui-react-core/esm2015/lib/SelectedItemsChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SelectedItemsChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class SelectedItemsChangedEventArgsDescription extends Description {
        constructor() {
            super();
            this.h = null;
            this.g = null;
            this.f = null;
        }
        get_type() {
            return "SelectedItemsChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get oldItems() {
            return this.h;
        }
        set oldItems(a) {
            this.h = a;
            this.e("OldItems");
        }
        get newItems() {
            return this.g;
        }
        set newItems(a) {
            this.g = a;
            this.e("NewItems");
        }
        get currentItems() {
            return this.f;
        }
        set currentItems(a) {
            this.f = a;
            this.e("CurrentItems");
        }
    }
    SelectedItemsChangedEventArgsDescription.$t = /*@__PURE__*/ markType(SelectedItemsChangedEventArgsDescription, 'SelectedItemsChangedEventArgsDescription', Description.$);
    return SelectedItemsChangedEventArgsDescription;
})();
