// igniteui-react-core/esm2015/lib/ShapeFilterRecordEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ShapeFilterRecordEventArgsDescription = /*@__PURE__*/ (() => {
    class ShapeFilterRecordEventArgsDescription extends Description {
        constructor() {
            super();
            this.f = null;
            this.h = false;
        }
        get_type() {
            return "ShapeFilterRecordEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get record() {
            return this.f;
        }
        set record(a) {
            this.f = a;
            this.e("Record");
        }
        get shouldInclude() {
            return this.h;
        }
        set shouldInclude(a) {
            this.h = a;
            this.e("ShouldInclude");
        }
    }
    ShapeFilterRecordEventArgsDescription.$t = /*@__PURE__*/ markType(ShapeFilterRecordEventArgsDescription, 'ShapeFilterRecordEventArgsDescription', Description.$);
    return ShapeFilterRecordEventArgsDescription;
})();
