// igniteui-react-core/esm2015/lib/StockPopulateEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let StockPopulateEventArgsDescription = /*@__PURE__*/ (() => {
    class StockPopulateEventArgsDescription extends Description {
        constructor() {
            super();
            this.g = null;
            this.f = null;
        }
        get_type() {
            return "StockPopulateEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get symbol() {
            return this.g;
        }
        set symbol(a) {
            this.g = a;
            this.e("Symbol");
        }
        get name() {
            return this.f;
        }
        set name(a) {
            this.f = a;
            this.e("Name");
        }
    }
    StockPopulateEventArgsDescription.$t = /*@__PURE__*/ markType(StockPopulateEventArgsDescription, 'StockPopulateEventArgsDescription', Description.$);
    return StockPopulateEventArgsDescription;
})();
