// igniteui-react-core/esm2015/lib/TemplateColumnDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ColumnDescription } from "./ColumnDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let TemplateColumnDescription = /*@__PURE__*/ (() => {
    class TemplateColumnDescription extends ColumnDescription {
        constructor() {
            super();
            this.ek = null;
            this.ej = null;
        }
        get templateRef() {
            return this.ek;
        }
        set templateRef(a) {
            this.ek = a;
            this.e("TemplateRef");
        }
        get_type() {
            return "TemplateColumn";
        }
        get cellUpdatingRef() {
            return this.ej;
        }
        set cellUpdatingRef(a) {
            this.ej = a;
            this.e("CellUpdatingRef");
        }
    }
    TemplateColumnDescription.$t = /*@__PURE__*/ markType(TemplateColumnDescription, 'TemplateColumnDescription', ColumnDescription.$);
    TemplateColumnDescription.__canMarshalByValue1 = true;
    return TemplateColumnDescription;
})();
