// igniteui-react-core/esm2015/lib/TransactionStateDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let TransactionStateDescription = /*@__PURE__*/ (() => {
    class TransactionStateDescription extends Description {
        constructor() {
            super();
            this.f = null;
            this.g = null;
            this.h = null;
            this.i = null;
        }
        get_type() {
            return "TransactionState";
        }
        get type() {
            return this.get_type();
        }
        get idRef() {
            return this.f;
        }
        set idRef(a) {
            this.f = a;
            this.e("IdRef");
        }
        get transactionType() {
            return this.g;
        }
        set transactionType(a) {
            this.g = a;
            this.e("TransactionType");
        }
        get valueRef() {
            return this.h;
        }
        set valueRef(a) {
            this.h = a;
            this.e("ValueRef");
        }
        get versionRef() {
            return this.i;
        }
        set versionRef(a) {
            this.i = a;
            this.e("VersionRef");
        }
    }
    TransactionStateDescription.$t = /*@__PURE__*/ markType(TransactionStateDescription, 'TransactionStateDescription', Description.$);
    return TransactionStateDescription;
})();
