// igniteui-react-core/esm2015/lib/ValueChangedEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ValueChangedEventArgsDescription = /*@__PURE__*/ (() => {
    class ValueChangedEventArgsDescription extends Description {
        constructor() {
            super();
            this.g = new Date();
            this.f = new Date();
        }
        get_type() {
            return "ValueChangedEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get oldValue() {
            return this.g;
        }
        set oldValue(a) {
            this.g = a;
            this.e("OldValue");
        }
        get newValue() {
            return this.f;
        }
        set newValue(a) {
            this.f = a;
            this.e("NewValue");
        }
    }
    ValueChangedEventArgsDescription.$t = /*@__PURE__*/ markType(ValueChangedEventArgsDescription, 'ValueChangedEventArgsDescription', Description.$);
    return ValueChangedEventArgsDescription;
})();
