// igniteui-react-core/esm2015/lib/VerticalStackedSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedSeriesBaseDescription } from "./StackedSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let VerticalStackedSeriesBaseDescription = /*@__PURE__*/ (() => {
    class VerticalStackedSeriesBaseDescription extends StackedSeriesBaseDescription {
        constructor() {
            super();
            this.c0 = null;
            this.c1 = null;
        }
        get_type() {
            return "VerticalStackedSeriesBase";
        }
        get xAxisRef() {
            return this.c0;
        }
        set xAxisRef(a) {
            this.c0 = a;
            this.e("XAxisRef");
        }
        get yAxisRef() {
            return this.c1;
        }
        set yAxisRef(a) {
            this.c1 = a;
            this.e("YAxisRef");
        }
    }
    VerticalStackedSeriesBaseDescription.$t = /*@__PURE__*/ markType(VerticalStackedSeriesBaseDescription, 'VerticalStackedSeriesBaseDescription', StackedSeriesBaseDescription.$);
    return VerticalStackedSeriesBaseDescription;
})();
